<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2020-09-15" />

<title>Estimating L and F</title>

<script src="site_libs/header-attrs-2.3/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_factors</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization">
    <span class="fa fa-gitlab"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Estimating <span class="math inline">\(L\)</span> and <span class="math inline">\(F\)</span></h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2020-09-15</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-09-25
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>sumstatFactors/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20190819code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20190819)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20190819code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20190819)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgitlabcomjeanmorrisongwassummaystatisticfactorizationtreef1b0a5a9fedeb914c0a3a0cd58b2684a51a052c8targetblankf1b0a5aa"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/tree/f1b0a5a9fedeb914c0a3a0cd58b2684a51a052c8" target="_blank">f1b0a5a</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgitlabcomjeanmorrisongwassummaystatisticfactorizationtreef1b0a5a9fedeb914c0a3a0cd58b2684a51a052c8targetblankf1b0a5aa" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/tree/f1b0a5a9fedeb914c0a3a0cd58b2684a51a052c8" target="_blank">f1b0a5a</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/figure/
    Ignored:    analysis/fixed_factors2_cache/
    Ignored:    analysis/fixed_factors_cache/
    Ignored:    analysis/metabo_flash_experiments_cache/
    Ignored:    analysis/pathway_factors_cache/
    Ignored:    analysis/simulations_cache/
    Ignored:    analysis/z_scores_cache/
    Ignored:    analysis_data/2020-06-10_ai/

Untracked files:
    Untracked:  R/est_R_pairwise.R
    Untracked:  analysis/k_greater_than_n.Rmd
    Untracked:  analysis/sample_corr2.Rmd
    Untracked:  analysis/simple_patterns_simulations.Rmd
    Untracked:  buttons.css
    Untracked:  bw_factors2.png
    Untracked:  code/hide_output.js
    Untracked:  k99_analysis.R
    Untracked:  k99_fit.RDS
    Untracked:  matrix_20.png

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/estimation.Rmd</code>) and HTML (<code>public/estimation.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://glcdn.githack.com/jean_morrison/gwas-summay-statistic-factorization/raw/f3643cebf98fa0fb754c5751a21447db21367701/public/estimation.html" target="_blank">f3643ce</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-09-25
</td>
<td>
Host with GitLab.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/284f75b7cf150993cb08439e8087c89ed3713505/analysis/estimation.Rmd" target="_blank">284f75b</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-09-16
</td>
<td>
rename and move stuff around for new structure
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>We introduced the model <a href="model.html">here</a>. In this doc we will discuss methods of estimating the <span class="math inline">\(L\)</span> and <span class="math inline">\(F\)</span> matrices.</p>
</div>
<div id="independent-gwas-independent-snps" class="section level2">
<h2>Independent GWAS, independent SNPs</h2>
<p>If all GWAS are conducted in different samples and all variants are independent, we can use FLASH out of the box. The only thing we have to worry about is if <span class="math inline">\(\Theta\)</span> is sparse rather than dense as assumed by FLASH. Some <a href="sample_cor_ld_sims.html">early simulations</a> suggest this may not be much of an issue. We also need to standardize our effect estimates. In September 2020, The basic code for applying FLASH to this problem is</p>
<pre class="r"><code>fit &lt;- flash.init(data=B_hat, S = S,  var.type=2) %&gt;%
          flash.add.greedy(Kmax = kmax, init.fn = init.fn.softImpute) %&gt;%
          flash.backfit() %&gt;%
          flash.nullcheck()</code></pre>
<p>Here <code>var.type=2</code> indicates that we want to estimate a different variance for every column of <span class="math inline">\(\Theta\)</span>. <code>B_hat</code> is the matrix of effect estimates and <code>S</code> is the matrix of standard errors.</p>
<p>Other variations of this are to use <span class="math inline">\(z\)</span>-scores or standardized versions of <span class="math inline">\(\hat{B}\)</span>. For simplicity, we assume that <span class="math inline">\(\hat{s}_{j,m}\)</span> is measured without error and that <span class="math inline">\(s^2_{j,m} = \frac{\sigma^2_{Y,m}}{N_m \sigma^2_{G,j}}\)</span> where <span class="math inline">\(\sigma^2_{G,j}\)</span> is the variance of variant <span class="math inline">\(j\)</span> and is constant across studies and <span class="math inline">\(\sigma^2_{Y,m}\)</span> is the variance of trait <span class="math inline">\(m\)</span>.</p>
<p>In this case <span class="math inline">\(s_{j,m}\)</span> factors into a variant component <span class="math inline">\(a_j\)</span> and trait component <span class="math inline">\(b_m\)</span>. Dividing by <span class="math inline">\(s_{j,m}\)</span> will divide rows or <span class="math inline">\(L\)</span> by <span class="math inline">\(a_j\)</span> and rows of <span class="math inline">\(F\)</span> by <span class="math inline">\(b_m\)</span>. Every element of <span class="math inline">\(\Theta\)</span> gets divided by <span class="math inline">\(a_j b_m\)</span> and the elements of <span class="math inline">\(E\)</span> are now IID <span class="math inline">\(N(0,1)\)</span>. Using FLASH this means we have changed our assumption about the distribution of <span class="math inline">\(L\)</span> and <span class="math inline">\(F\)</span>. Rather than assuming that effects of variants on factors are IID from some distribution (in the case above a spike and slab distribution), we are assuming that effects scaled by SNP variance are IID and that factor effects scaled by sample size over trait variance are IID. Except for the scaling by sample size these assumptions are fairly reasonable, or at least as reasonable as the assumptions they are replacing. Therefore, we might consider a “standardized” effect of <span class="math inline">\(\hat{\beta}_{j,m} \sigma_{G,j}/\sigma_{Y,m}\)</span> or more computable using our data, <span class="math inline">\(\frac{\hat{\beta}_{j,m}}{\hat{s}_{j,m}\sqrt{N_m}} = \frac{\hat{z}_{j,m}}{\sqrt{N_m}}\)</span>. In this document, I will refer to standardized effect etsimates as <span class="math inline">\(\tilde{\beta}_{j,m}\)</span>.</p>
<p>To fit with standardized effects, assuming <code>N</code> is a length <span class="math inline">\(M\)</span> vector of sample sizes:</p>
<pre class="r"><code>B_tilde = t( (1/sqrt(N)) *t(B_hat/S))
S_tilde = t( (1/sqrt(N)) * t(matrix(1, nrow=n_var, ncol=n_triat)))
fit &lt;- flash.init(data=B_tilde, S = S_tilde,  var.type=2) %&gt;%
          flash.add.greedy(Kmax = kmax, init.fn = init.fn.softImpute) %&gt;%
          flash.backfit() %&gt;%
          flash.nullcheck()</code></pre>
</div>
<div id="overlapping-samples-independent-snps" class="section level2">
<h2>Overlapping samples, independent SNPs</h2>
<p>To data with overlapping samples between traits we need to correct for it. There are two elements of this problem: 1) estimating the correlation of rows of <span class="math inline">\(E\)</span> and 2) correcting for it in the fit.</p>
<p>I will address the second part first. I have considered two alternative methods for correcting for a known covariance in the rows of <span class="math inline">\(E\)</span>. Both work fairly well in simulations and have different drawbacks in real data.</p>
<div id="eigenvector-transformation" class="section level3">
<h3>Eigenvector transformation</h3>
<p>The pros of this method are that it is (of the two I will talk about) easier to fit and seems to perform better in simulations. The cons are that it can only be used with complete data. It also applies sparsity assumptions to the wrong matrix. First suppose that every element of <span class="math inline">\(E\)</span> has the same variance. In this section we will assume we are working with standardized effect esimtates <span class="math inline">\(\tilde{\beta}_{j,m}\)</span>. This also works with <span class="math inline">\(z\)</span>-scores but not with unstandardized effects (with unstandardized effects we would need to transform effects for each SNP separately). Our model is</p>
<p><span class="math display">\[
\tilde{B} = LF^\top + \Theta + E
\]</span> where the covariance of each row of <span class="math inline">\(E\)</span> is <span class="math inline">\(\Sigma = SRS\)</span> where <span class="math inline">\(S\)</span> is <span class="math inline">\(M\times M\)</span> diagonal with diagonal elements <span class="math inline">\(1/\sqrt{N_m}\)</span> and <span class="math inline">\(R\)</span> is a known correlation matrix. Let $R = VDV^$ be the eigendecomposition of <span class="math inline">\(R\)</span>. Then</p>
<p><span class="math display">\[
\tilde{B}V = L (V^\top F)^\top + \Theta V + E_2 
\]</span> where the elements of <span class="math inline">\(E_2\)</span> are independent with <span class="math inline">\(Var((E_2)_{j,m}) = d_{m}/N_m\)</span> where <span class="math inline">\(d_m\)</span> is the <span class="math inline">\(m\)</span>th diagonal element of <span class="math inline">\(D\)</span>. We can infer <span class="math inline">\(L\)</span> and <span class="math inline">\(F\)</span> by first transforming <span class="math inline">\(\tilde{B}\)</span> and then back transforming the estimated factors. Unfortunately we are now assuming that <span class="math inline">\(V^\top F\)</span> is sparse rather than assuming that <span class="math inline">\(F\)</span> is sparse. In simulations this doesn’t seem to make a huge difference. However, the inability to use incomplete SNPs may significantly impact real data analysis. To fit in FLASH we use</p>
<pre class="r"><code>fit_ev &lt;- function(B_hat, S_hat, N, R, kmax=100, zero_thresh = 1e-15){
  
    n_var &lt;- nrow(B_hat)
    n_trait &lt;- ncol(B_hat)
    stopifnot(nrow(S_hat) == n_var &amp; ncol(S_hat) == n_trait)
    stopifnot(length(N) == n_trait)
    
    R_eig &lt;- eigen(R)
    d &lt;- R_eig$values
    V &lt;- R_eig$vectors
    
    R_eig$values[abs(v) &lt; zero_thresh] &lt;- 0
    if(any(R_eig$values &lt; 0))stop(&quot;R is not psd&quot;)
    
    B_tilde = t( (1/sqrt(N)) *t(B_hat/S))
    S_tilde = t( (1/sqrt(N)) * t(matrix(1, nrow=n_var, ncol=n_triat)))
    
    B_tilde_tilde &lt;- B_tilde %*% V
    S_tilde_tilde &lt;- t( (sqrt(d)/sqrt(N)) * t(matrix(1, nrow=n_var, ncol=n_trait)))
    
    fit &lt;- flash.init(data=B_tilde_tilde, S = S_tilde_tilde,  var.type=2) %&gt;%
                      flash.add.greedy(Kmax = kmax, init.fn = init.fn.softImpute) %&gt;%
                      flash.backfit() %&gt;%
                      flash.nullcheck();
    
    if(fit$n.factors &gt; 0){
      F_hat &lt;- V %*% fit$loadings.pm[[2]]
      L_hat &lt;- fit$loadings.pm[[1]]
      B_hat &lt;- fitted(fit) %*% t(V)
    }else{
      B_hat &lt;- F_hat  &lt;- L_hat &lt;- NULL;
    }
  ret &lt;- list(fit=fit, B_hat = B_hat, L_hat = L_hat, F_hat = F_hat)
  return(ret)
}</code></pre>
</div>
<div id="addition-of-fixed-factors" class="section level3">
<h3>Addition of fixed factors</h3>
<p>This method is harder to fit but can make use of variants that are only present in a subset of studies. It also maintains the same sparsity assumptions as the original problem. This strategy was explored previously by Matthew and Jason <a href="https://willwerscheid.github.io/FLASHvestigations/arbitraryV.html">here</a>.</p>
<p>Let <span class="math inline">\(\lambda_{min}\)</span> be the smallest eigenvalue of <span class="math inline">\(\Sigma\)</span> and let <span class="math inline">\(\Sigma - \lambda_{min}I = V DV^\top\)</span> (note that I have re-used <span class="math inline">\(V\)</span> and <span class="math inline">\(D\)</span> to mean something different than they meant in the previous section). Then</p>
<p><span class="math display">\[
E = AW^\top + E_1
\]</span></p>
<p>where elements of <span class="math inline">\(E_1\)</span> are iid N<span class="math inline">\((0, \lambda_{min})\)</span>, the elements of <span class="math inline">\(A\)</span> are iid <span class="math inline">\(N(0, 1)\)</span> and <span class="math inline">\(W = V D^{1/2}\)</span>. So theoretically, we should be able to fit the model by adding in the known fixed factors <span class="math inline">\(W\)</span> and fixing the prior for the columns of <span class="math inline">\(A\)</span>. In practice this is a bit hard. My first attempts failed and I found that the order of fitting each part mattered quite a bit. The function below works pretty well in simulations:</p>
<pre class="r"><code>fit_ff &lt;- function(B_hat, S_hat, N, R, kmax=100){
  
  n_var &lt;- nrow(B_hat)
  n_trait &lt;- ncol(B_hat)
  stopifnot(nrow(S_hat) == n_var &amp; ncol(S_hat) == n_trait)
  stopifnot(length(N) == n_trait)
  
  Sigma &lt;- diag(sqrt(N)) %*% R %*% diag(sqrt(N))  
  lambda_min &lt;- eigen(Sigma) %&gt;% 
                with(., min(values))
  Sig_new &lt;- Sigma - lambda_min*diag(rep(1, ntrait))
  
  B_tilde = t( (1/sqrt(N)) *t(B_hat/S))
  S_tilde = t( (1/sqrt(N)) * t(matrix(1, nrow=n_var, ncol=n_triat)))
  
  if(all(Sig_new ==0)){
    fit &lt;- NULL
    B_hat &lt;- NULL
    F_hat &lt;- NULL
  }else{
    Z &lt;- with(dat, beta_hat/se_beta_hat)
    Sig_eig &lt;- eigen(Sig_new)
    V &lt;- Sig_eig$vectors[, -n_trait]
    W &lt;- V %*% sqrt(Sig_eig$values[-n_trait])
    
    # randomly initialize A
    A_rand &lt;- matrix(rnorm(n=nvar*(ntrait-1)), nrow=nvar, ncol=(ntrait-1))

    #First add some greedy factors but don&#39;t backfit
    fit &lt;-  flash.init(B_tilde, S = sqrt(lambda_min), var.type = 2) %&gt;%
            flash.add.greedy(Kmax = ntrait, init.fn = init.fn.default )
    #Next add in fixed factorsl. Use sequential mode for backfit
    n &lt;- fit$n.factors;
    fit &lt;- fit %&gt;%
           flash.init.factors(., EF = list(A_rand, W), prior.family = prior.normal(scale= 1)) %&gt;%
           flash.fix.loadings(., kset = n + 1:(ntrait-1), mode=2) %&gt;%
           flash.backfit(method = &quot;sequential&quot;)
    
    F_hat &lt;- fit$loadings.pm[[2]][,1:n]
    fixed_ix &lt;- n + (1:(ntrait-1))
    B_hat &lt;- fitted(fit) - with(fit, loadings.pm[[1]][, fixed_ix]%*%diag(loadings.scale[fixed_ix])%*% t(loadings.pm[[2]][, fixed_ix]))
  }  
}</code></pre>
</div>
<div id="estimating-sigma" class="section level3">
<h3>Estimating <span class="math inline">\(\Sigma\)</span></h3>
<p>So far I have only used a straight forward approach of estimating <span class="math inline">\(R\)</span>, the correlation of rows of <span class="math inline">\(E\)</span> empirically with $<em>{m, m} = Cor(</em>{, m}, _{, m^}) $. In real data I estimate each element of <span class="math inline">\(R\)</span> using slightly different variants due to missingness. I found that limiting only to variants non-missing yields too few variants to accurately estimate the correlations. Theoretically, this no longer guarantees that the the resulting matrix is positive semi-definite but in practice it has not been a problem.</p>
<p>I did some experiments limiting the variants used to estimate <span class="math inline">\(R\)</span> based on a <span class="math inline">\(p\)</span>-value threshold. Ideally, if we knew a subset of variants that were associated with zero rows of <span class="math inline">\(L\)</span>, it would be best to use these variants. If <span class="math inline">\(\Theta\)</span> is sparse, these variants should be those with zero marginal effect so we can try to approximate such a set by setting a <span class="math inline">\(p\)</span>-value threshold and only taking variants above that threshold for all traits (or both traits if estimating pair-wise). In simulations, this always performed worse than using all variants. What is best in real data is an open question.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5      rstudioapi_0.11 whisker_0.4     knitr_1.29     
 [5] magrittr_1.5    R6_2.4.1        rlang_0.4.6     stringr_1.4.0  
 [9] tools_4.0.2     xfun_0.15       git2r_0.27.1    htmltools_0.5.0
[13] ellipsis_0.3.1  rprojroot_1.3-2 yaml_2.2.1      digest_0.6.25  
[17] tibble_3.0.2    lifecycle_0.2.0 crayon_1.3.4    later_1.1.0.1  
[21] vctrs_0.3.1     promises_1.1.1  fs_1.4.2        glue_1.4.1     
[25] evaluate_0.14   rmarkdown_2.3   stringi_1.4.6   compiler_4.0.2 
[29] pillar_1.4.4    backports_1.1.8 httpuv_1.5.4    pkgconfig_2.0.3</code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
