<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2020-11-11" />

<title>Pipeline</title>

<script src="site_libs/header-attrs-2.3/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_factors</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization">
    <span class="fa fa-gitlab"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Pipeline</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2020-11-11</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-12-01
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>sumstatFactors/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20190819code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20190819)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20190819code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20190819)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgitlabcomjeanmorrisongwassummaystatisticfactorizationtreed43d7570d18a55f5ef8a0f3696177232bd094c1btargetblankd43d757a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/tree/d43d7570d18a55f5ef8a0f3696177232bd094c1b" target="_blank">d43d757</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgitlabcomjeanmorrisongwassummaystatisticfactorizationtreed43d7570d18a55f5ef8a0f3696177232bd094c1btargetblankd43d757a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/tree/d43d7570d18a55f5ef8a0f3696177232bd094c1b" target="_blank">d43d757</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/fixed_factors2_cache/
    Ignored:    analysis/fixed_factors_cache/
    Ignored:    analysis/metabo_flash_experiments_cache/
    Ignored:    analysis/pathway_factors_cache/
    Ignored:    analysis/simulations_cache/
    Ignored:    analysis/z_scores_cache/
    Ignored:    analysis_data/2020-06-10_ai/
    Ignored:    analysis_data/2020-10-02_metabolites/
    Ignored:    analysis_data/simple_patterns/
    Ignored:    analysis_data/sims_medium_k/
    Ignored:    pipeline/.snakemake/auxiliary/
    Ignored:    pipeline/.snakemake/conda-archive/
    Ignored:    pipeline/.snakemake/conda/
    Ignored:    pipeline/.snakemake/incomplete/
    Ignored:    pipeline/.snakemake/locks/
    Ignored:    pipeline/.snakemake/metadata/
    Ignored:    pipeline/.snakemake/shadow/
    Ignored:    pipeline/.snakemake/singularity/

Untracked files:
    Untracked:  analysis/sample_corr2.Rmd
    Untracked:  buttons.css
    Untracked:  bw_factors2.png
    Untracked:  code/hide_output.js
    Untracked:  k99_analysis.R
    Untracked:  k99_fit.RDS
    Untracked:  matrix_20.png

Unstaged changes:
    Modified:   R/est_L.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/pipeline.rmd</code>) and HTML (<code>public/pipeline.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/d43d7570d18a55f5ef8a0f3696177232bd094c1b/analysis/pipeline.rmd" target="_blank">d43d757</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-12-01
</td>
<td>
wflow_publish(“analysis/pipeline.rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://glcdn.githack.com/jean_morrison/gwas-summay-statistic-factorization/raw/998527e950024a6b2b1a0e121a140b4f0fe896b8/public/pipeline.html" target="_blank">998527e</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-12-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/b8b5a5b959b10d215d044744bf076aedc2a42dfd/analysis/pipeline.rmd" target="_blank">b8b5a5b</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-12-01
</td>
<td>
wflow_publish(“analysis/pipeline.rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://glcdn.githack.com/jean_morrison/gwas-summay-statistic-factorization/raw/713b7556c50643fa82f589c8162a2a3402118ae0/public/pipeline.html" target="_blank">713b755</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-11-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/823fbfa6570578030235316a896ea23a52e70a7a/analysis/pipeline.rmd" target="_blank">823fbfa</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-11-24
</td>
<td>
wflow_publish(“analysis/pipeline.rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://glcdn.githack.com/jean_morrison/gwas-summay-statistic-factorization/raw/3016fa129188f77885413e3f1c4ae83b7a659d84/public/pipeline.html" target="_blank">3016fa1</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-11-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://glcdn.githack.com/jean_morrison/gwas-summay-statistic-factorization/raw/3b34e6e74d7d993143d07cd403a21fea5ec85461/public/pipeline.html" target="_blank">3b34e6e</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-11-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/ebbcb4991b9df1f0d850776fba200b5b8fefb869/analysis/pipeline.rmd" target="_blank">ebbcb49</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-11-11
</td>
<td>
analysis/index.Rmd
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/blob/a3434536d74af077ca370c1e0d086176fb4189ad/analysis/pipeline.rmd" target="_blank">a343453</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-11-11
</td>
<td>
add pipeline
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>I have written a <a href="https://snakemake.readthedocs.io/en/stable/index.html">Snakemake</a> pipeline for analyzing data. This is very much in progress so changes may happen frequently or there may be new versions. I will update this document as things change.</p>
<p>To download the pipeline skeleton you can use a function in the R package I have made for this project, the <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/-/blob/master/README.md"><code>sumstatFactors</code> package</a>.</p>
<pre class="r"><code>sumstatFactors::pipeline_init()</code></pre>
<p>TODO: This is not implemented yet. For now, copy the directory <code>/project2/xinhe/jean/gwas_factors/data_analysis/2020-11-11_test_pipeline/pipeline</code> on RCC.</p>
<p>This will download files:</p>
<ul>
<li><code>Snakefile</code>: The pipeline file</li>
<li><code>config.yaml</code>: Config file for specifying analysis options</li>
<li><code>cluster.yaml</code>: Cluster file. You won’t have to change this.</li>
<li><code>run-snakemake.sh</code>: Submit command for running the Snakemake. If you are on RCC you won’t have to change this unless you need to change accounts.</li>
<li>A directory of R code. This is called by the Snakefile.</li>
<li>An example csv file for specifying data.</li>
</ul>
</div>
<div id="requirements" class="section level2">
<h2>Requirements</h2>
<p>To run the pipeline you need Snakemake installed. You also need the following R packages:</p>
<ul>
<li><code>sumstatFactors</code>: This is the package I made for developing this project. You can find instructions <a href="https://gitlab.com/jean_morrison/gwas-summay-statistic-factorization/-/blob/master/README.md">here</a>.</li>
<li>Tidyverse packages: <code>stringr</code>, <code>readr</code>, <code>deplyr</code>, <code>tidyr</code>, <code>purrr</code></li>
<li><code>gwasvcf</code>, <code>VariantAnnotation</code> for dealing with vcf files. See <a href="https://mrcieu.github.io/gwasvcf/">here</a>.</li>
<li><code>ieugwasr</code>: For LD pruning with plink. See <a href="https://mrcieu.github.io/ieugwasr/">here</a>.<br />
</li>
<li><code>LaplacesDemon</code>: I only use one function from here, <code>as.symmetric.matrix</code> there may be a workaround later that doesn’t require this.</li>
</ul>
</div>
<div id="how-it-works" class="section level2">
<h2>How it works</h2>
<p>The Snakefile reads options from <code>config.yaml</code> and runs the specified analysis. To specify your analysis you will modify <code>config.yaml</code> and provide a csv file specified there. Ideally you will not have to change anything else. You might have to modify <code>run-snakemake.sh</code> if you have different options for submitting your jobs (like if you want to use a different partition or a different PI account on RCC).</p>
<p>This is a good time to go look at the config file. The config file has three sections. I will walk through each part here. Currently there is no validation of parameter inputs in the config file so read the following sections carefully so you can provide appropriate inputs. In the future we can implement validation.</p>
<div id="input" class="section level3">
<h3>Input</h3>
<p>This section specifies the data files that contain your input GWAS summary statistics. The only parameter in the input section is <code>sum_stats</code>. This gives a csv that contains information about the GWAS you want to analyze. The input section of the yaml file looks like this:</p>
<pre><code>input:
    sum_stats: &quot;my_gwas_files.csv&quot;</code></pre>
<p>There is an example csv file downloaded when you download the pipeline. There is one row per study. The csv file tells the location and format of the raw gwas summary statistics. There are two types of files that you can provide, flat files with any delimiter (like csv, tsv, space separated etc), or vcf files downloaded from the IEU open gwas project. You can have some studies of each type if necessary. If the file is a flat file, it is ok if it is gzipped. vcf files must end in <code>.vcf.gz</code>. The two types of files are entered a bit differently in the spread sheet, so read the instructions carefully below.</p>
<p>The csv file should have the following fields in any order:</p>
<ul>
<li>name: A unique string for each trait.</li>
<li>raw_data_path: Path to the raw summary statistics (usually as downloaded is fine)</li>
<li>snp, A1, A2, beta_hat, se, chrom, pos, p_value, sample_size: If the raw data is a flat file (for example tsv or csv), then these are column names in the raw data for columns giving SNP rs number, effect allele, non-effect allele, effect estimate, standard error of effect estimate, chromosome, position, p-value, and sample size. p-value and sample size can be listed as NA if that column is missing. The other columns are required. If the data is a vcf file, these fields can all be given as NA as they will be ignored.</li>
<li>delimiter: This can be “space”, “tab” “comma” or any character. It indicates the delimiter of the file.</li>
<li>effect_is_or: Yes or No. If Yes this indicates that beta_hat is an odds ratio (rather than log odds ratio) and it will be transformed.</li>
<li>pub_sample_size: Sample size provided in the corresponding publication. If sample_size is not present in the summary statistics file, this sample size will be used instead.</li>
</ul>
<p>The csv file should contain one entry for each study you want to include. Don’t list any files you don’t want to include. In some rare cases you may have to create a modified version of the data before using the pipeline. Sometimes the fields of rs number or standard error are omitted. At some point I may add some functionality to deal with these cases, however, the IEU open GWAS project is also collecting studies and standardizing them so this may stop being an issue we run into frequently.</p>
</div>
<div id="analysis" class="section level3">
<h3>Analysis</h3>
<p>This section specifies how you want the analysis to be run. There are several parameters and it will look like this:</p>
<pre><code>analysis:
    r2_plink: 0.01
    ref_path: &quot;/project2/compbio/LD/plink_reference/EUR&quot;
    R_pthresh: 0.2
    kmax: 100
    norm_type: [&quot;ss&quot;, &quot;z&quot;]
    min_nonmissing: [123]
    method: [&quot;plain&quot;, &quot;ev&quot;, &quot;ff&quot;]
    flash_pthresh: [1, 0.01]
    seed: 2
    est_L: True</code></pre>
<p>Here is a description of each parameter:</p>
<ul>
<li><code>r2_plink</code>: r^2 threshold for LD pruning</li>
<li><code>ref_path</code>: Path to LD reference files used by plink</li>
<li><code>R_pthresh</code>: p-value threshold for computing <span class="math inline">\(R\)</span>, the correlation matrix for rows of <span class="math inline">\(E\)</span>. For a particular pair of traits, the function will use all variants with <span class="math inline">\(p\)</span>-value greater than the threshold for both traits.</li>
<li><code>kmax</code>: Maximum number of factors for FLASH</li>
<li><code>norm_type</code>: The type of statistics you want to use. This should be a python syntax list of one or two items. Each item can be either <code>ss</code> or <code>z</code>. <code>ss</code> indicates normalized effect estimates (i.e. z-score divided by square root of sample size) while <code>z</code> indicates <code>z</code>-scores. Setting this option as <code>["ss"]</code> would run only one analysis using normalized effect estimates. Setting it as <code>["ss", "z"]</code> will run both.</li>
<li><code>min_nonmissing</code>: Variants will be retained if they have summary statistics for at least this number of traits. This parameter should be a python syntax list of integers.</li>
<li><code>method</code>: A python syntax list of methods to run. <code>plain</code> indicates not adjusting for sample overlap. <code>ev</code> indicates using the eigenvector transformation method, <code>ff</code> indicates using the fixed factor method. Currently if <code>ev</code> is uses with <code>min_nonmissing</code> less than the full number of traits you will get an error from the flash step of the pipeline. This is a future item to improve about the pipeline.</li>
<li><code>flash_pthresh</code>: A python syntax list of numbers between 0 an 1 giving the p-value threshold for FLASH. FLASH will only use SNPs that have a p-value less than this threshold for at least one trait.</li>
<li><code>seed</code>: A random seed</li>
<li><code>est_L</code>: Either True or False. Whether to run the step of estimating <span class="math inline">\(\hat{L}\)</span> and standard errors. This generates fairly big files so it might be best to start with False and then only compute <span class="math inline">\(\hat{L}\)</span> for analyses you are interested in.</li>
</ul>
<p>Most likely, you will only need to modify <code>min_nonmissing</code>, <code>method</code>, <code>norm_type</code>. <code>flash_pthresh</code>, and <code>est_L</code>. You may modify seed to see how robust the results are. The other parameters are at reasonable defaults. If you use data with non-european ancestry, you may want to change the LD reference data.</p>
</div>
<div id="output" class="section level3">
<h3>Output</h3>
<p>This section specifies where to put output files and what to call them. It looks like this:</p>
<pre><code>out:
    data_dir: &quot;data/&quot;
    formatted_gwas_dir: &quot;/project2/compbio/gwas_summary_statistics/standard_formats/ss_factors/&quot;
    output_dir: &quot;results/&quot;
    prefix: &quot;my-prefix_&quot;</code></pre>
<ul>
<li><code>data_dir</code>: A directory where interim data files will be stored.</li>
<li><code>formatted_gwas_dir</code>: A directory where formatted GWAS data will be stored. This is separate because these files aren’t analsis specific and you may want to re-use them.</li>
<li><code>output_dir</code>: A directory where results will be stored.<br />
</li>
<li><code>prefix</code>: An analysis specific prefix that will be pre-pended to all files produced.</li>
</ul>
<p>You should make sure that the directory names end in “/” and that the prefix name ends in "_".</p>
</div>
</div>
<div id="how-to-run-it" class="section level2">
<h2>How to run it</h2>
<p>It is always good to check that the pipeline will run and will do what you think it will before you actually run this. To do this, run Snakemake with the dry run option (<code>-n</code>). I like to output the results to a file so I can look at them later:</p>
<pre><code>snakemake -n &gt; plan.txt</code></pre>
<p>This will either give you an error if something is wrong (for example maybe one of your files is missing) or it will print out a list of all the jobs it will run. If this runs without error and looks as you expect you are ready to run the pipeline.</p>
<p>I prefer to submit the pipeline from an <code>sinteractive</code> session on a compute node. This makes it easy to kill in the middle if I want to by terminating the session. If you do this, make sure to log on to start the session with the <code>screen</code> command. This will let you “detatch” the session so that it keeps running even when the window is closed. You can use a command like</p>
<pre><code>screen sinteractive --account pi-mstephens --partition mstephens  --mem 5G --time 4-00:00:00</code></pre>
<p>This starts an <code>sinteractive</code> session on the <code>mstephens</code> partition that will run for 4 days (unless you kill it early) and will be allotted 5G of memory.</p>
<p>Next run Snakemake using</p>
<pre><code>nohup ./run-snakemake.sh &amp; </code></pre>
<p>The <code>nohup</code> and <code>&amp;</code> parts make sure that the command keeps running when the window is closed. You might need to give <code>run-snakemake.sh</code> executable permissions using <code>chmod a+x run-snakemake.sh</code>.</p>
<p>When this is submitted, Snakemake will run and will take care of submitting jobs to the cluster. Log files will be written to a subdirectory called <code>log</code>. You can see output as it runs in <code>nohup.out</code> (unless you redirect to another output file using <code>nohup ./run-snakemake.sh &gt; myfile.out &amp;</code>.</p>
<p>You can detatch your screen session using Ctrl-d. </p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.6.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5      rstudioapi_0.12 whisker_0.4     knitr_1.30     
 [5] magrittr_1.5    R6_2.5.0        rlang_0.4.8     stringr_1.4.0  
 [9] tools_4.0.3     xfun_0.18       git2r_0.27.1    htmltools_0.5.0
[13] ellipsis_0.3.1  rprojroot_1.3-2 yaml_2.2.1      digest_0.6.27  
[17] tibble_3.0.4    lifecycle_0.2.0 crayon_1.3.4    later_1.1.0.1  
[21] vctrs_0.3.4     promises_1.1.1  fs_1.5.0        glue_1.4.2     
[25] evaluate_0.14   rmarkdown_2.3   stringi_1.5.3   compiler_4.0.3 
[29] pillar_1.4.6    backports_1.2.0 httpuv_1.5.4    pkgconfig_2.0.3</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
