<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2020-06-18" />

<title>Immune Mediated Diseases</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_factors</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jean997/ss_factors">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Immune Mediated Diseases</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2020-06-18</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-06-18
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>sumstatFactors/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0.9000). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20190819code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20190819)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20190819code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20190819)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstree1972eb05a51b6019c620d3c868047fba1c43839ftargetblank1972eb0a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jean997/sumstatFactors/tree/1972eb05a51b6019c620d3c868047fba1c43839f" target="_blank">1972eb0</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstree1972eb05a51b6019c620d3c868047fba1c43839ftargetblank1972eb0a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/fixed_factors2_cache/
    Ignored:    analysis/fixed_factors_cache/
    Ignored:    analysis/metabo_flash_experiments_cache/
    Ignored:    analysis/pathway_factors_cache/
    Ignored:    analysis/simulations_cache/
    Ignored:    analysis/z_scores_cache/
    Ignored:    analysis_data/2020-06-10_ai/

Untracked files:
    Untracked:  R/est_R_pairwise.R
    Untracked:  analysis/2020-04-28-meeting_notes.Rmd
    Untracked:  analysis/literature.Rmd
    Untracked:  analysis/sample_corr2.Rmd
    Untracked:  analysis/sumstat_factors.bib
    Untracked:  buttons.css
    Untracked:  code/flashier_compare.R
    Untracked:  code/flashier_compare_jason.R
    Untracked:  code/hide_output.js
    Untracked:  code/mask_flashier.R
    Untracked:  code/mask_flashr.R
    Untracked:  covariance_scratch.R
    Untracked:  docs/figure/ld_etc.Rmd/
    Untracked:  docs/figure/sample_corr2.Rmd/
    Untracked:  for_jason/
    Untracked:  k99_analysis.R
    Untracked:  k99_fit.RDS
    Untracked:  man/sim_sumstats_lf.Rd
    Untracked:  simulate_1.rds
    Untracked:  simulate_2.rds

Unstaged changes:
    Modified:   man/mask_and_fit.Rd
    Modified:   man/sumstatFactors.Rd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/1972eb05a51b6019c620d3c868047fba1c43839f/analysis/imd.Rmd" target="_blank">1972eb0</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
<td>
wflow_publish(“analysis/imd.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/sumstatFactors/bca6b19c77de59e123c876848ed9229f53ab317c/docs/imd.html" target="_blank">bca6b19</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/60fd564a786eb6c4d8843a6e88d2d6575f64b566/analysis/imd.Rmd" target="_blank">60fd564</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
<td>
wflow_publish(“analysis/imd.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This is an exploration of an analysis of a set of immune mediated diseases exploring some analysis choices.</p>
<table>
<thead>
<tr class="header">
<th>Disease</th>
<th>Abbrv</th>
<th>Cases</th>
<th>Controls</th>
<th>SNPS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Lupus</td>
<td>SLE</td>
<td>7219</td>
<td>15991</td>
<td>6,762,880</td>
</tr>
<tr class="even">
<td>Rheumatoid Arthritis</td>
<td>RA</td>
<td>29880</td>
<td>73758</td>
<td>8,240,308</td>
</tr>
<tr class="odd">
<td>Chrons Disease</td>
<td>CD</td>
<td>22575</td>
<td>46693</td>
<td>9,458,484</td>
</tr>
<tr class="even">
<td>Inflammatory Bowel Disease</td>
<td>IBD</td>
<td>42950</td>
<td>53536</td>
<td>9,934,092</td>
</tr>
<tr class="odd">
<td>Ulcerative Colitis</td>
<td>UC</td>
<td>20417</td>
<td>52230</td>
<td>9,554,059</td>
</tr>
<tr class="even">
<td>Asthma</td>
<td>Asth</td>
<td>23948</td>
<td>118538</td>
<td>2,001,281</td>
</tr>
<tr class="odd">
<td>Primary Schlerosing Cholangitis</td>
<td>PSC</td>
<td>4796</td>
<td>19955</td>
<td>6,805,391</td>
</tr>
<tr class="even">
<td>Allergy</td>
<td>Allg</td>
<td>180129</td>
<td>180709</td>
<td>7,112,974</td>
</tr>
<tr class="odd">
<td>Ankylosing Spondilitis</td>
<td>Anksp</td>
<td>10619</td>
<td>15145</td>
<td>103,085</td>
</tr>
<tr class="even">
<td>Celiac</td>
<td>Cel</td>
<td>4533</td>
<td>10750</td>
<td>521,934</td>
</tr>
<tr class="odd">
<td>Multiple Sclerosis</td>
<td>MS</td>
<td>14498</td>
<td>24091</td>
<td>136,408</td>
</tr>
<tr class="even">
<td>Psoriasis</td>
<td>PSO</td>
<td>10588</td>
<td>22806</td>
<td>141,528</td>
</tr>
<tr class="odd">
<td>Primary Biliary Cirrhosis</td>
<td>PBC</td>
<td>2764</td>
<td>10475</td>
<td>1,047,436</td>
</tr>
<tr class="even">
<td>hildhood Ashtma</td>
<td>Ch. Asth (Ukb)</td>
<td>9676</td>
<td>347481</td>
<td>7,038,206</td>
</tr>
<tr class="odd">
<td>Ashtma (both)</td>
<td>Asth (Ukb)</td>
<td>46802</td>
<td>347481</td>
<td>7,038,206</td>
</tr>
<tr class="even">
<td>Adult onset asthma</td>
<td>Ad. Asth (Ukb)</td>
<td>22296</td>
<td>347481</td>
<td>7,038,206</td>
</tr>
<tr class="odd">
<td>Latend autoimmune diabetes</td>
<td>LADA</td>
<td>2634</td>
<td>5947</td>
<td>4,293,816</td>
</tr>
<tr class="even">
<td>Vitiligo</td>
<td>Vit</td>
<td>2853</td>
<td>37405</td>
<td>7,490,324</td>
</tr>
<tr class="odd">
<td>Type 1 Diabetes</td>
<td>T1D</td>
<td>6683</td>
<td>12173</td>
<td>107,814</td>
</tr>
</tbody>
</table>
<p>There are a few challenges with these data. The fist is that some but not all studies have overlapping samples. UC, CD, and IBD are all measured in the same samples as are three of the asthma phenotypes (the three marked as (Ukb) above). A second challenge is that some the othe studies, Anksp, MS, PSO, and T1D only have published summary statistics for immunochip SNPs which is a much smaller set than the usual genome-wide study. In all analyes variants are pruned for LD</p>
</div>
<div id="non-missing-variants-various-sample-overlap-corrections" class="section level2">
<h2>Non-missing variants, various sample overlap corrections</h2>
<p>The first set of analyses uses only variants that are non-missing in all studies and uses several methods to correct for sample overlap. we consider three estimates of <span class="math inline">\(R\)</span>, the residual correlation in rows of <span class="math inline">\(E\)</span>.</p>
<ol style="list-style-type: decimal">
<li>Identity (no sample overlap)</li>
<li><span class="math inline">\(R\)</span> estimated with all SNPs in all studies (assumes possible sample overlap between any pair of studies) (“all”).</li>
<li>Correlation estimated only between studies with known overlapping samples (“overlap”).</li>
</ol>
<p>We consider two methods of correcting for sample overlap, the eigenvector transformation and fixed factors. In this case, since studies are mostly non-overlapping this analysis tells us how much power/accuracy we lose if we correct for sample overlap when it is not present. For comparison we will also look at the SVD of all variants.</p>
<pre><code>Warning: replacing previous import &#39;intervals::reduce&#39; by &#39;purrr::reduce&#39;
when loading &#39;sumstatFactors&#39;</code></pre>
<pre><code>
Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    filter, lag</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, setdiff, setequal, union</code></pre>
<pre><code>
Attaching package: &#39;gridExtra&#39;</code></pre>
<pre><code>The following object is masked from &#39;package:dplyr&#39;:

    combine</code></pre>
<p>Without correcting for sample overlap: <img src="figure/imd.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-2-1">
Past versions of unnamed-chunk-2-1.png
</button>
</p>
<div id="fig-unnamed-chunk-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/bca6b19c77de59e123c876848ed9229f53ab317c/docs/figure/imd.Rmd/unnamed-chunk-2-1.png" target="_blank">bca6b19</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A few interesting patterns: factors 1 and 2 capture IBD/UC/CD and ashthma/allergy clusters respectively. These are probably the strongest patterns in these data. Factor 5 captures differences between UC and CD plus some overlap of UC and PSC and RA which can be concommitant. Factor 6 is unique to childhood (not adult onset) asthma and is also shared with PBC, SLE, and RA.</p>
<p>Correcting for sample overlap across all studies: <img src="figure/imd.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/bca6b19c77de59e123c876848ed9229f53ab317c/docs/figure/imd.Rmd/unnamed-chunk-3-1.png" target="_blank">bca6b19</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Results using the eigenvector transformation are “muddy” while the fixed factor estimator is more sparse and has similar patterns to the using <span class="math inline">\(R=I\)</span>. Factor 12 captures sharing between T1D and LADA which we would expect to see.</p>
<p>Correcting for sample overlap in only some studies: <img src="figure/imd.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/bca6b19c77de59e123c876848ed9229f53ab317c/docs/figure/imd.Rmd/unnamed-chunk-4-1.png" target="_blank">bca6b19</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Here are two SVD decompositions. In the first we just take the SVD of the entire matrix of <span class="math inline">\(z\)</span>-scores. In the second, <span class="math inline">\(z\)</span>-scores corresponding to <span class="math inline">\(p\)</span>-values greater than 0.05 are set to zero. These are extremely similar. <img src="figure/imd.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/bca6b19c77de59e123c876848ed9229f53ab317c/docs/figure/imd.Rmd/unnamed-chunk-5-1.png" target="_blank">bca6b19</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/imd.Rmd/unnamed-chunk-5-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-2">
Past versions of unnamed-chunk-5-2.png
</button>
</p>
<div id="fig-unnamed-chunk-5-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/bca6b19c77de59e123c876848ed9229f53ab317c/docs/figure/imd.Rmd/unnamed-chunk-5-2.png" target="_blank">bca6b19</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-06-19
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="variants-present-in-at-least-two-studies" class="section level2">
<h2>Variants present in at least two studies</h2>
<p>If we allow missing data we can use many more variants. This analysis has 168,142 variants rather than 3,991. We can no longer use the eigenvector transformation. Both with and without correction, these fits are “less neat” than the non-missing data fits. This may because the imuno chip traits have very high (90%) missingness. To explore further, I will analyze only the non-imunochip traits.</p>
<p><img src="figure/imd.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/imd.Rmd/unnamed-chunk-6-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="non-missing-variants-non-immunochip-studies" class="section level2">
<h2>Non-missing variants, non immunochip studies</h2>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] gridExtra_2.3             dplyr_0.8.3              
[3] ggplot2_3.2.1             sumstatFactors_0.0.0.9000

loaded via a namespace (and not attached):
 [1] tidyselect_0.2.5     xfun_0.9             ashr_2.2-47         
 [4] purrr_0.3.3          reshape2_1.4.3       lattice_0.20-38     
 [7] colorspace_1.4-1     vctrs_0.2.0          htmltools_0.3.6     
[10] yaml_2.2.0           rlang_0.4.2          mixsqp_0.3-43       
[13] pillar_1.4.2         glue_1.3.1           withr_2.1.2         
[16] DBI_1.0.0            lifecycle_0.1.0      plyr_1.8.5          
[19] stringr_1.4.0        munsell_0.5.0        gtable_0.3.0        
[22] workflowr_1.4.0.9000 flashier_0.2.4       evaluate_0.14       
[25] labeling_0.3         knitr_1.24           invgamma_1.1        
[28] irlba_2.3.3          parallel_3.6.3       Rcpp_1.0.4.6        
[31] readr_1.3.1          backports_1.1.5      scales_1.1.0        
[34] truncnorm_1.0-8      farver_2.0.1         fs_1.3.1            
[37] RMySQL_0.10.17       hms_0.5.2            digest_0.6.23       
[40] stringi_1.4.3        ebnm_0.1-24          grid_3.6.3          
[43] rprojroot_1.3-2      tools_3.6.3          magrittr_1.5        
[46] lazyeval_0.2.2       tibble_2.1.3         crayon_1.3.4        
[49] whisker_0.4          pkgconfig_2.0.3      zeallot_0.1.0       
[52] Matrix_1.2-17        SQUAREM_2020.2       assertthat_0.2.1    
[55] rmarkdown_1.15       R6_2.4.1             intervals_0.15.1    
[58] git2r_0.26.1         compiler_3.6.3      </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
