<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2020-09-03" />

<title>Model</title>

<script src="site_libs/header-attrs-2.3/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_factors</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jean997/ss_factors">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Model</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2020-09-03</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-09-16
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>sumstatFactors/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstree2b876239da50fb8326d9c924376693344b9388e1targetblank2b87623a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jean997/sumstatFactors/tree/2b876239da50fb8326d9c924376693344b9388e1" target="_blank">2b87623</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstree2b876239da50fb8326d9c924376693344b9388e1targetblank2b87623a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jean997/sumstatFactors/tree/2b876239da50fb8326d9c924376693344b9388e1" target="_blank">2b87623</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/fixed_factors2_cache/
    Ignored:    analysis/fixed_factors_cache/
    Ignored:    analysis/metabo_flash_experiments_cache/
    Ignored:    analysis/pathway_factors_cache/
    Ignored:    analysis/simulations_cache/
    Ignored:    analysis/z_scores_cache/
    Ignored:    analysis_data/2020-06-10_ai/

Untracked files:
    Untracked:  R/est_R_pairwise.R
    Untracked:  analysis/k_greater_than_n.Rmd
    Untracked:  analysis/sample_corr2.Rmd
    Untracked:  analysis/simple_patterns_simulations.Rmd
    Untracked:  buttons.css
    Untracked:  bw_factors2.png
    Untracked:  code/hide_output.js
    Untracked:  k99_analysis.R
    Untracked:  k99_fit.RDS
    Untracked:  matrix_20.png

Unstaged changes:
    Modified:   R/sim_sumstats_lf.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/model.Rmd</code>) and HTML (<code>docs/model.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/284f75b7cf150993cb08439e8087c89ed3713505/analysis/model.Rmd" target="_blank">284f75b</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-09-16
</td>
<td>
rename and move stuff around for new structure
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jean997/sumstatFactors/7a2b85f3ea3443a2538c35f74960c5220b95fdd0/docs/model.html" target="_blank">7a2b85f</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-09-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/a4f0dce2d90b54e19357858ed1c83f337bb5def3/analysis/model.Rmd" target="_blank">a4f0dce</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-09-15
</td>
<td>
wflow_publish(c(“analysis/model.Rmd”, “analysis/intro.Rmd”, “analysis/index.Rmd”))
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="data-generating-model" class="section level2">
<h2>Data Generating Model</h2>
<p>Here we show the details of our model for marginal GWAS summary statistics. Throughout <span class="math inline">\(m\)</span> indexes traits, <span class="math inline">\(j\)</span> indexes variants, <span class="math inline">\(k\)</span> indexes factors, and <span class="math inline">\(i\)</span> indexes individuals. We are condiserding <span class="math inline">\(M\)</span> total traits and <span class="math inline">\(J\)</span> total variants. For each trait-variant pair, we observe the marginal association of variant <span class="math inline">\(j\)</span> with trait <span class="math inline">\(m\)</span> but we don’t get to observe the individual level data. We denote the marginal effect estimate <span class="math inline">\(\hat{\beta}_{m,j}\)</span> and the corresponding standard error <span class="math inline">\(\hat{s}_{m,j}\)</span>.</p>
<p>We assume that these effect estimates originated from a GWAS, so within a trait effect estimates for all variants were computed using the same sample. Across traits, samples may overlap entirely, partially or not at all.</p>
<p>We model <span class="math inline">\(K\)</span> (<span class="math inline">\(K\)</span> is unknown) unobserved underlying “factors”. Each factor, <span class="math inline">\(F_{k}\)</span> has a linear effect of <span class="math inline">\(f_{m,k}\)</span> on trait <span class="math inline">\(m\)</span>. Each variant <span class="math inline">\(G_j\)</span> has a linear effect if <span class="math inline">\(l_{j,k}\)</span> on factor <span class="math inline">\(k\)</span>. Vairant <span class="math inline">\(j\)</span> may have an additional linear effect on trait <span class="math inline">\(m\)</span>, <span class="math inline">\(\theta_{j,m}\)</span> that is not mediated by factors. (See picture below)</p>
<p>Below is a list of assumptions we will use in this document. I have marked with a star assumptions that we can weaken or remove later.</p>
<ol style="list-style-type: decimal">
<li>Every GWAS has appropriately corrected for population structure. Thuse every marginal effect measures either a causal effect of the variant on the trait or causal effects of nearby linked variants. *</li>
<li>All variants are independent (no LD). *</li>
<li>Effects of unobserved factors on traits are linear.</li>
<li>Effects of variants on factors are linear.</li>
<li>Effects of factors on traits and variants on factors are sparse (many are equal to zero).</li>
<li>For a given <span class="math inline">\(m\)</span>, <span class="math inline">\(\theta_{j,m} \sim N(0, \sigma_{\theta, m}^2)\)</span>. <span class="math inline">\(\theta_{j,m}\)</span> and <span class="math inline">\(\theta_{j^\prime, m^\prime}\)</span> are independent if <span class="math inline">\(j \neq j^\prime\)</span> or <span class="math inline">\(m \neq m^\prime\)</span>.</li>
<li>The elements of <span class="math inline">\(L\)</span> are IID with <span class="math inline">\(Var(l_{j,k}) = \sigma^2_{l,k}\)</span>.</li>
<li>The columns of <span class="math inline">\(F\)</span> have been scaled so that <span class="math inline">\(\sum_{m=1}^M f_{m,k}^2 = 1\)</span>. Here we are thinking of <span class="math inline">\(F\)</span> as fixed and <span class="math inline">\(L\)</span> as random.</li>
</ol>
</div>
<div id="gwas-with-no-overlapping-samples" class="section level2">
<h2>GWAS with no overlapping samples</h2>
<p>This is the simplest case. Using our previous assumptions, for a single individual <span class="math inline">\(i\)</span> in the study of trait <span class="math inline">\(m\)</span>, we can write</p>
<p><span class="math display">\[
Y_{m,i} = \sum_{k=1}^{K}f_{mk}F_{k,i} + \sum_{j=1}^{J} \theta_{j,m}G^{(m)}_{j,i} + \epsilon_{m,i}
\]</span></p>
<p>where <span class="math inline">\(\epsilon_{m,i}\)</span> is any non-genetic component of <span class="math inline">\(Y_{m,i}\)</span> that is not mediated through the factors. The superscript <span class="math inline">\((m)\)</span> on <span class="math inline">\(G^{(m)}_{j,i}\)</span> reminds us that these are the genotypes measured in study <span class="math inline">\(m\)</span>.</p>
<p>Each GWAS produces <span class="math inline">\(\beta_{j, m}\)</span>, the association of variant <span class="math inline">\(j\)</span> with trait <span class="math inline">\(m\)</span>. Using our assumption that all effect variants are independent,</p>
<p><span class="math display">\[
\beta_{j,m} = \sum_{k=1}^{K}l_{jk}f_{mk} + \theta_{j,m}
\]</span> If each variant explains a small proportion of the overall trait variance, then the variance of <span class="math inline">\(\hat{\beta}_{j,m}\)</span>, the OLS estimate of <span class="math inline">\(\beta_{j,m}\)</span> is approximately <span class="math inline">\(\frac{\sigma^2_{Ym}}{N_m \sigma^2_{Gj}}\)</span> where <span class="math inline">\(\sigma^2_{Ym}\)</span> is the variance of trait <span class="math inline">\(m\)</span> and <span class="math inline">\(\sigma^2_{Gj}\)</span> os the variance of variant <span class="math inline">\(j\)</span>, or <span class="math inline">\(2 f_j (1-f_j)\)</span> where <span class="math inline">\(f_j\)</span> is the allele frequency of variant <span class="math inline">\(j\)</span>. It will be convenient later to assume that both the traits and the variants have been standardized so that the variance of <span class="math inline">\(\hat{\beta}_{j,m}\)</span> is about <span class="math inline">\(1/N_m\)</span>.</p>
<p>Let <span class="math inline">\(\hat{B}\)</span> be the <span class="math inline">\(J \times M\)</span> matrix of marginal effect estimates and <span class="math inline">\(B\)</span> be the <span class="math inline">\(J \times M\)</span> matrix of true marginl associations. Then</p>
<p><span class="math display">\[
\hat{B} = B + E \\
 = L F^\top + \Theta + E
\]</span> where <span class="math inline">\(L\)</span> is the <span class="math inline">\(J\times K\)</span> matrix of SNP effects, <span class="math inline">\(F\)</span> is the <span class="math inline">\(M \times K\)</span> matrix of factor effects on traits and <span class="math inline">\(\Theta\)</span> is the <span class="math inline">\(J\times M\)</span> matrix of vairant effects not mediated by factors. Because all studies are independent and all variants are independent, the elements of <span class="math inline">\(E\)</span> are independent.</p>
<p>This is precisely the problem that FLASH is designed to solve, with additional assumptions that <span class="math inline">\(L\)</span> and <span class="math inline">\(F\)</span> are sparse and that <span class="math inline">\(\Theta\)</span> is dense. FLASH can estimate a different distribution for each column of <span class="math inline">\(L\)</span> and <span class="math inline">\(F\)</span> and each column of <span class="math inline">\(\Theta\)</span>.</p>
</div>
<div id="fully-overlapping-samples" class="section level2">
<h2>Fully overlapping samples</h2>
<p>If samples fully overlap between the GWAS of all <span class="math inline">\(M\)</span> traits, the covariance of a row of <span class="math inline">\(E\)</span> will be <span class="math inline">\(\frac{1}{N}\Sigma\)</span> where <span class="math inline">\(\Sigma\)</span> is the <span class="math inline">\(M \times M\)</span> covariance of <span class="math inline">\(Y_1, \dots, Y_m\)</span>. In practice, we will not know <span class="math inline">\(\Sigma\)</span> and will have to estimate it. However for the sake of generating realistic data and more fully characterizing the problem, it is worth considering the factors tha contribute to <span class="math inline">\(\Sigma\)</span>.</p>
<p>We can decompose <span class="math inline">\(\Sigma\)</span> into three components:</p>
<p><span class="math display">\[
\Sigma = \Sigma_G + \Sigma_{FE} + \Sigma_E
\]</span></p>
<p><span class="math inline">\(\Sigma_G\)</span> is the genetic covariance while <span class="math inline">\(\Sigma_{FE}\)</span> and <span class="math inline">\(\Sigma_E\)</span> are non-genetic. We assume that shared factors may have both genetic and non-genetic effects and <span class="math inline">\(\Sigma_{FE}\)</span> represents the non-genetic (environmental) covariance due to shared factors. <span class="math inline">\(\Sigma_E\)</span> is environmental covariance not mediated by the heritble factors. We can write out expressions for each of these components:</p>
<p><span class="math display">\[
(\Sigma_G)_{m, m^\prime} = \sum_{j=1}^J E\left[\left(\sum_{k=1}^K l_{jk}f_{mk} + \theta_{jm}\right)\left(\sum_{k=1}^K l_{jk}f_{m^\prime k} + \theta_{jm^\prime}\right)\right]  \\
 = \sum_{j=1}^JE\left[\sum_{k =1}^{K} l_{j,k}^2 f_{m,k} f_{m^\prime, k}\right]  + E\left[\theta_{j,m}\theta_{j,m^\prime}\right]   \\
 = J \sum_{k=1}^K \sigma_{lk}^2  f_{m,k} f_{m^\prime, k} + JCov(\theta_m, \theta_{m^\prime})
 \]</span> Where <span class="math inline">\(\sigma_{lk}^2 = Var(l_{j,k})\)</span> and <span class="math inline">\(Cov(\theta_m, \theta_{m^\prime}) = \sigma_{\theta,m}^2\)</span> if <span class="math inline">\(m = m^\prime\)</span> and 0 otherwise. Here we are thinking of <span class="math inline">\(F\)</span> as fixed and <span class="math inline">\(L\)</span> as random. In matrix form</p>
<p><span class="math display">\[
\Sigma_G = J (FV_LF^\top + V_\theta)
\]</span> where <span class="math inline">\(V_\theta\)</span> is a diagonal <span class="math inline">\(M\times M\)</span> matrix who’s elements are equal to <span class="math inline">\(\sigma^2_{\theta, m}\)</span> and <span class="math inline">\(V_L\)</span> is diagonal <span class="math inline">\(K\times K\)</span> with elements <span class="math inline">\(\sigma^2_{l,k}\)</span>.</p>
<p><span class="math inline">\(\Sigma_{FE} = FV_{F}F^\top\)</span> where <span class="math inline">\(V_F\)</span> is a <span class="math inline">\(K \times K\)</span> diagonal matrix who’s elements are <span class="math inline">\(\sigma^2_{FE,k}\)</span>.</p>
<p><span class="math inline">\(\Sigma_E\)</span> is the non-genetic, non-factor covariance and could be anything.</p>
<p>An important thing to note here is that <span class="math inline">\(\Sigma_G\)</span> and <span class="math inline">\(\Sigma_{FE}\)</span> are both have components that are <span class="math inline">\(FDF^\top\)</span> for some diagonal matrix D. This means we can re-write <span class="math inline">\(E\)</span> as</p>
<p><span class="math display">\[
E = E_1 + E_2\\
= \tilde{L}F^\top + E_2
\]</span></p>
<p>where <span class="math inline">\(\tilde{L}\)</span> is a <span class="math inline">\(J\times K\)</span> matrix with independent elements <span class="math inline">\(\tilde{L}_{j,k} \sim N(0, \omega_{k}^2)\)</span> and the rows of <span class="math inline">\(E_2\)</span> have correlation due only to environmental covariance not mediated by factors.</p>
<p>An important difference between <span class="math inline">\(L\)</span> and <span class="math inline">\(\tilde{L}\)</span> is that <span class="math inline">\(L\)</span> is sparse while <span class="math inline">\(\tilde{L}\)</span> is dense. However, if <span class="math inline">\(L\)</span> is not sparse enough it can be difficult to distinguish the two in estimation.</p>
</div>
<div id="partially-overlapping-samples" class="section level2">
<h2>Partially overlapping samples</h2>
<p>In practice we are likely to have varying degrees of sample overlap across different samples. This might be in blocks, for example a subset of traits are all from the same study (e.g. UK Biobank) while the remaining traits are independent. There might also be pairs of traits that have</p>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
