<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2019-09-06" />

<title>Fixed factors from pathways</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_factors</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jean997/ss_factors">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Fixed factors from pathways</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2019-09-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-09-06
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>sumstatFactors/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0.9000). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20190819code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20190819)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20190819code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20190819)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
pathway1
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory <code>pathway_factors_cache</code> and re-run the analysis. To have workflowr automatically delete the cache directory prior to building the file, set <code>delete_cache = TRUE</code> when running <code>wflow_build()</code> or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstreef1d6d49dfabeccf432838cac57993cd2041f8467targetblankf1d6d49a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jean997/sumstatFactors/tree/f1d6d49dfabeccf432838cac57993cd2041f8467" target="_blank">f1d6d49</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstreef1d6d49dfabeccf432838cac57993cd2041f8467targetblankf1d6d49a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/fixed_factors_cache/
    Ignored:    analysis/metabo_flash_experiments_cache/
    Ignored:    analysis/pathway_factors_cache/

Untracked files:
    Untracked:  R/snp_to_pathway.R
    Untracked:  analysis_data/.~lock.metabo2_gwas.csv#
    Untracked:  analysis_data/HACER-enhancers.txt
    Untracked:  analysis_data/HACER_README.txt
    Untracked:  analysis_data/bc_gwas_mats.RDS
    Untracked:  analysis_data/eqtl_genes.RDS
    Untracked:  analysis_data/flashier_res2019-09-03.RDS
    Untracked:  analysis_data/metabo2_gwas.csv
    Untracked:  analysis_data/metabo3_order1__flashier__by_col.RDS
    Untracked:  analysis_data/metabo3_order1__flashier__by_row.RDS
    Untracked:  analysis_data/metabo3_order1__flashier__constant.RDS
    Untracked:  analysis_data/metabo3_order1__flashier__kronecker.RDS
    Untracked:  analysis_data/metabo3_order1__flashier__noisy_bycol.RDS
    Untracked:  analysis_data/metabo3_order1__flashier__noisy_byrow.RDS
    Untracked:  analysis_data/metabo3_order1__flashier__noisy_constant.RDS
    Untracked:  analysis_data/metabo3_order1__flashier__zero.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__by_col.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__by_row.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__constant.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__kronecker.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__noisy_bycol.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__noisy_byrow.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__noisy_constant.RDS
    Untracked:  analysis_data/metabo3_order1__from_data__zero.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__by_col.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__by_row.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__constant.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__kronecker.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__noisy_bycol.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__noisy_byrow.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__noisy_constant.RDS
    Untracked:  analysis_data/metabo3_order1__soft_impute__zero.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__by_col.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__by_row.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__constant.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__kronecker.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__noisy_bycol.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__noisy_byrow.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__noisy_constant.RDS
    Untracked:  analysis_data/metabo3_order1_mask__flashier__zero.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__by_col.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__by_row.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__constant.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__kronecker.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__noisy_bycol.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__noisy_byrow.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__noisy_constant.RDS
    Untracked:  analysis_data/metabo3_order1_mask__from_data__zero.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__by_col.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__by_row.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__constant.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__kronecker.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__noisy_bycol.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__noisy_byrow.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__noisy_constant.RDS
    Untracked:  analysis_data/metabo3_order1_mask__soft_impute__zero.RDS
    Untracked:  analysis_data/metabo_gwas.csv
    Untracked:  analysis_data/pathway_loadings_data.RDS
    Untracked:  buttons.css
    Untracked:  code/flashier_compare.R
    Untracked:  code/hide_output.js
    Untracked:  docs/figure/fixed_factors.Rmd/
    Untracked:  man/annotate_snp_loc.Rd
    Untracked:  man/genes_within.Rd
    Untracked:  man/get_eqtl_genes.Rd
    Untracked:  snp_annot.RDS

Unstaged changes:
    Modified:   NAMESPACE
    Modified:   R/package.R
    Modified:   analysis/basic_flash_analysis.Rmd
    Modified:   analysis/metabo_flash_experiments.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/f1d6d49dfabeccf432838cac57993cd2041f8467/analysis/pathway_factors.Rmd" target="_blank">f1d6d49</a>
</td>
<td>
Jean Morrison
</td>
<td>
2019-09-07
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/metabo_flash_experiments_order1.Rmd”,
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="compiling-pathway-information" class="section level2">
<h2>Compiling Pathway information</h2>
<pre><code>Warning: replacing previous import &#39;intervals::reduce&#39; by &#39;purrr::reduce&#39;
when loading &#39;sumstatFactors&#39;</code></pre>
<pre class="r"><code>mats &lt;- readRDS(&quot;analysis_data/metabo3_gwas_mats_order1.RDS&quot;)
mats$beta_hat[is.na(mats$se_hat)] &lt;- NA
mats$se_hat[is.na(mats$beta_hat)] &lt;- Inf

traits &lt;- str_split(mats$traits, &quot;/&quot;) %&gt;% map(., 2) %&gt;% 
          unlist(.) %&gt;%
          str_replace(., &quot;.top_summary_statistics.tsv.gz&quot;, &quot;&quot;) %&gt;%
          str_replace(., &quot;metabo3_&quot;, &quot;&quot;)

ntrait &lt;- length(mats$traits)
nsnp &lt;- length(mats$snps)</code></pre>
<p>Here the goal is to define some loadings that can only have non-zero values for SNPs within a single pathway. In order to do this we need to first map SNPs to genes and genes to pathways which is a bit tricky. My first attempt will take a broad approach. I will use the following rules to associate SNPs with genes:</p>
<ol style="list-style-type: decimal">
<li>The SNP is a cis eQTL for the gene in GTEx v8 in any tissue</li>
<li>The SNP is within X Kb of the start or end of the gene.</li>
<li>The SNP is in an enhancer that is associated with the gene in either FANTOM5 or 4DGenome</li>
</ol>
<p>The last bit of information is obtained from HACER (Human ACtive Enhancer to interpret Regulatory variants, <a href="http://bioinfo.vanderbilt.edu/AE/HACER/index.html" class="uri">http://bioinfo.vanderbilt.edu/AE/HACER/index.html</a> )</p>
<p>The first step is to get all of these annotations.</p>
<pre class="r"><code># Get SNP annotation information
snp_annot &lt;- annotate_snp_loc(mats$snps)
head(snp_annot)
# A tibble: 6 x 7
  name       chrom       pos ref   alt   func    has_func
  &lt;chr&gt;      &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;lgl&gt;   
1 rs1000940  17      5283252 C     T     intron  FALSE   
2 rs10011174 4     153495515 A     G     unknown FALSE   
3 rs10016041 4     140117746 A     G     unknown FALSE   
4 rs10019888 4      26062990 A     G     unknown FALSE   
5 rs10023335 4      77358987 C     T     intron  FALSE   
6 rs1002607  9     113858933 C     T     unknown FALSE   </code></pre>
<p>Note that we only have annotation information for 4389 of the 4393 SNPs.</p>
<p>The eQTL information I extracted on rcc using the following code from the <code>/project2/compbio/external_public_supp/GTEx/V8/GTEx_Analysis_v8_eQTL</code> directory.</p>
<pre class="r"><code>gtex_files &lt;- list.files(&quot;.&quot;, &quot;.v8.signif_variant_gene_pairs.chrpos.tsv.gz&quot;)
tissues &lt;- str_replace(gtex_files, &quot;.v8.signif_variant_gene_pairs.chrpos.tsv.gz&quot;, &quot;&quot;)
eqtl_genes &lt;- get_eqtl_genes(snp_annot,gtex_files, tissues)</code></pre>
<p>This produced a data frame listing SNP-Gene pairs and in which tissue the SNP is an eQTL for the gene.</p>
<pre class="r"><code>eqtl_genes &lt;- readRDS(&quot;analysis_data/eqtl_genes.RDS&quot;)
dim(eqtl_genes)
[1] 767   6
head(eqtl_genes)
# A tibble: 6 x 6
  chrom    pos name      chr   gene_id            tissue              
  &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;              &lt;chr&gt;               
1 11    204228 rs1045454 chr11 ENSG00000177963.14 Adipose_Subcutaneous
2 11    204228 rs1045454 chr11 ENSG00000142082.14 Adipose_Subcutaneous
3 11    204228 rs1045454 chr11 ENSG00000185627.17 Adipose_Subcutaneous
4 11    204228 rs1045454 chr11 ENSG00000277290.1  Adipose_Subcutaneous
5 11    494662 rs10540   chr11 ENSG00000174915.11 Adipose_Subcutaneous
6 11    494662 rs10540   chr11 ENSG00000023191.16 Adipose_Subcutaneous</code></pre>
<p>This have us a table 767 SNP-Gene pairs. However, many of them are duplicated becaseu they are detected in multiple tissues.</p>
<pre class="r"><code>eqtl_genes &lt;- distinct(eqtl_genes, chrom, pos, name, gene_id)</code></pre>
<p>Reducing to only unique pairs leaves 109 SNP-Gene pairs.</p>
<p>From the table of enhancers we extract all SNP-Gene pairs where the SNP is in an enhancer that targets the gene in either FANTOM5 or 4DGenome</p>
<pre class="r"><code>hacer_df &lt;- read_tsv(&quot;analysis_data/HACER-enhancers.txt&quot;)
Parsed with column specification:
cols(
  .default = col_character(),
  start = col_double(),
  end = col_double(),
  center = col_number(),
  PMID = col_number(),
  distance = col_double(),
  `Normalized-count` = col_double(),
  density = col_double(),
  VISTA = col_double(),
  Ensembl = col_double(),
  ENCODE = col_double(),
  ChromHMM = col_double()
)
See spec(...) for full column specifications.
enhancer_genes &lt;- get_enhancer_genes(snp_annot, hacer_df)
head(enhancer_genes)
# A tibble: 6 x 7
  snp         Enhancer_ID         chr       start       end source  gene   
  &lt;chr&gt;       &lt;chr&gt;               &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  
1 rs114693598 AE_hg19_A172_170262 chr1  161500608 161500899 FANTOM5 USF1   
2 rs114693598 AE_hg19_A172_170262 chr1  161500608 161500899 FANTOM5 ARHGAP…
3 rs114693598 AE_hg19_A172_170262 chr1  161500608 161500899 FANTOM5 FCER1G 
4 rs114693598 AE_hg19_A172_170262 chr1  161500608 161500899 FANTOM5 FCGR2A 
5 rs114693598 AE_hg19_A172_170262 chr1  161500608 161500899 FANTOM5 ATF6   
6 rs114693598 AE_hg19_A172_170262 chr1  161500608 161500899 FANTOM5 HSPA6  </code></pre>
<p>This yields 6359 SNP-Gene pairs but again there are overlaps because often the same SNP falls in many annotated enhancers in different cell types that are all linked with the same gene. There are only 187 distinct pairings. So eQTL and enhancer information have together not provided very much information for linking SNPs the genes and then to pathways.</p>
<p>We can also link SNPs to gene by distance. We will use the following rules</p>
<ol style="list-style-type: decimal">
<li>A SNP is linked to its closest gene if it is within 500kb. In the case that a SNP is inside of multiple genes it is linked to all.</li>
<li>Additionally, a SNP is linked to any gene with a TSS within 2kb</li>
<li>If the SNP is more than 500kb from any gene it is not linked to a gene</li>
</ol>
<pre class="r"><code>gene_info &lt;- retrieve_genes()
Joining, by = &quot;ucsc_name&quot;
Joining, by = &quot;ucsc_name&quot;
near_genes &lt;- nearest_gene(snp_annot, gene_info) %&gt;%
              filter(gene_dist &lt;=500000)
Joining, by = &quot;chrom&quot;
tss2kb_genes &lt;- genes_within(snp_annot, gene_info, dist=2000)</code></pre>
<p>Now we have all the info, time to combine into a single data frame</p>
<pre class="r"><code>eqtl_genes &lt;- eqtl_genes %&gt;% 
              distinct(name, gene_id) %&gt;% 
              mutate(gene_id = gsub(&quot;\\..*&quot; ,&quot;&quot;, gene_id),
                     method = &quot;eQTL&quot;) %&gt;%
              rename(snp = name)
enhancer_genes &lt;- enhancer_genes %&gt;%
                  distinct(snp, gene, source) %&gt;%
                  rename(geneSymbol = gene, method=source) %&gt;%
                  left_join(., gene_info) %&gt;%
                  select(snp, ensembl_gene_stable_id, method) %&gt;%
                  rename(gene_id = ensembl_gene_stable_id)
Joining, by = &quot;geneSymbol&quot;
near_genes &lt;- near_genes %&gt;%
              distinct(name, ensembl_gene_stable_id) %&gt;%
              rename(gene_id = ensembl_gene_stable_id, snp=name) %&gt;%
              mutate(method=&quot;closest&quot;)

 tss2kb_genes &lt;- tss2kb_genes  %&gt;% 
                 rename(gene_id = gene) %&gt;%
                 mutate(method=&quot;Within_2kb&quot;)
 
 all_genes &lt;- bind_rows(eqtl_genes, enhancer_genes, near_genes, tss2kb_genes)
 with(all_genes, table(method))
method
   closest       eQTL    FANTOM5 Within_2kb 
      4364        109        187        264 </code></pre>
</div>
<div id="which-go-ids-to-use" class="section level2">
<h2>Which GO IDs to use</h2>
<p>Most of our SNP-Gene pairs come from the closest gene method. There are 4651 distinct pairings. Now we can associate genes with GO Ids.</p>
<pre class="r"><code>go_ids &lt;- retrieve_go_ids(unique(all_genes$gene_id)) %&gt;% 
          filter(go_id != &quot;&quot;)</code></pre>
<p>There are 9888 unique GO IDs associated with genes in our list. We can summarize them by number of gene.</p>
<pre class="r"><code>x &lt;- go_ids %&gt;% group_by(go_id, name_1006) %&gt;% summarize(n = length(unique(ensembl_gene_id)))
summary(x$n)
    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
   1.000    1.000    2.000    6.284    4.000 1758.000 </code></pre>
<p>Almost half, 45%, of pathways have only one gene associated with SNPs in our list. However, some GO IDs are very general and are associated with a large number of genes.</p>
<pre class="r"><code>x[order(-x$n),][1:10,]
# A tibble: 10 x 3
# Groups:   go_id [10]
   go_id      name_1006                          n
   &lt;chr&gt;      &lt;chr&gt;                          &lt;int&gt;
 1 GO:0005515 protein binding                 1758
 2 GO:0016020 membrane                        1179
 3 GO:0005737 cytoplasm                       1091
 4 GO:0005634 nucleus                         1077
 5 GO:0005829 cytosol                          816
 6 GO:0005886 plasma membrane                  744
 7 GO:0016021 integral component of membrane   728
 8 GO:0046872 metal ion binding                608
 9 GO:0005654 nucleoplasm                      555
10 GO:0003677 DNA binding                      411</code></pre>
<p>Very general pathways won’t be informative and pathways with almost no genes probably won’t be useful.</p>
<p>Summarizing GO IDs by number of SNPs associated with genes in them:</p>
<pre class="r"><code>all_genes &lt;- distinct(all_genes, snp, gene_id)
gids &lt;- unique(go_ids$go_id)
snps &lt;- map(gids, function(id){
  genes &lt;- filter(go_ids, go_id == id) %&gt;% with(., unique(ensembl_gene_id))
  snps &lt;-  filter(all_genes, gene_id %in% genes) %&gt;% with(., unique(snp))
})
nsnps &lt;- sapply(snps, length)
summary(nsnps)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.00    1.00    3.00   10.24    7.00 2627.00 </code></pre>
<p>gives a similar pattern. We can limit the pathways we use by the number of SNPs, the number of genes, or both. For the first try I will use pathways with at least 10 SNPs, at least 10 genes, and fewer than 200 genes. The rational being that pathways with very few SNPs are unlikely to be useful and pathways with a large number of genes are likely to be very gneral pathways like “nucleus”.</p>
<pre class="r"><code>go_id_summ &lt;- tibble(go_id = gids, nsnps = nsnps) %&gt;% full_join(x, by=&quot;go_id&quot;)
my_go_ids &lt;- filter(go_id_summ, n &lt;= 200 &amp; nsnps &gt;= 10 &amp; n &gt;=10) %&gt;% with(., go_id)</code></pre>
<p>This leaves 991 pathways.</p>
<pre class="r"><code>npathway &lt;- length(my_go_ids)

ix &lt;- with(go_id_summ, which(n &lt;= 200 &amp; nsnps &gt;= 10 &amp; n &gt;=10))
my_snps &lt;- snps[ix]
LL &lt;- matrix(0, nrow=nsnp, ncol=npathway)
fixl &lt;- matrix(TRUE, nrow=nsnp, ncol=npathway)
for(i in seq(npathway)){
  s &lt;- my_snps[[i]]
  ix &lt;- which(mats$snps %in% s)
  LL[ix, i] &lt;- 1
  fixl[ix, i] &lt;- FALSE
}</code></pre>
<pre class="r"><code>ldf &lt;- flash_get_ldf(f6, drop_zero_factors = FALSE)
plot_factors(ldf$f, traits, factor_names = )</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] reshape2_1.4.3            forcats_0.4.0            
 [3] stringr_1.4.0             dplyr_0.8.3              
 [5] purrr_0.3.2               readr_1.3.1              
 [7] tidyr_0.8.3               tibble_2.1.3             
 [9] ggplot2_3.2.1             tidyverse_1.2.1          
[11] sumstatFactors_0.0.0.9000 flashr_0.6-6             

loaded via a namespace (and not attached):
 [1] httr_1.4.1           RMySQL_0.10.17       jsonlite_1.6        
 [4] foreach_1.4.7        modelr_0.1.5         assertthat_0.2.1    
 [7] mixsqp_0.1-97        cellranger_1.1.0     yaml_2.2.0          
[10] ebnm_0.1-24          pillar_1.4.2         backports_1.1.4     
[13] lattice_0.20-38      glue_1.3.1           digest_0.6.20       
[16] rvest_0.3.4          colorspace_1.4-1     htmltools_0.3.6     
[19] Matrix_1.2-17        plyr_1.8.4           pkgconfig_2.0.2     
[22] broom_0.5.2          haven_2.1.1          scales_1.0.0        
[25] intervals_0.15.1     whisker_0.4          git2r_0.26.1        
[28] generics_0.0.2       withr_2.1.2          ashr_2.2-32         
[31] lazyeval_0.2.2       cli_1.1.0            magrittr_1.5        
[34] crayon_1.3.4         readxl_1.3.1         evaluate_0.14       
[37] fansi_0.4.0          fs_1.3.1             doParallel_1.0.15   
[40] nlme_3.1-141         MASS_7.3-51.4        xml2_1.2.2          
[43] truncnorm_1.0-8      tools_3.6.1          hms_0.5.1           
[46] softImpute_1.4       munsell_0.5.0        flashier_0.1.16     
[49] compiler_3.6.1       rlang_0.4.0          grid_3.6.1          
[52] iterators_1.0.12     rstudioapi_0.10      rmarkdown_1.15      
[55] gtable_0.3.0         codetools_0.2-16     DBI_1.0.0           
[58] R6_2.4.0             lubridate_1.7.4      knitr_1.24          
[61] utf8_1.1.4           zeallot_0.1.0        workflowr_1.4.0.9000
[64] rprojroot_1.3-2      stringi_1.4.3        pscl_1.5.2          
[67] parallel_3.6.1       SQUAREM_2017.10-1    Rcpp_1.0.2          
[70] vctrs_0.2.0          dbplyr_1.4.2         tidyselect_0.2.5    
[73] xfun_0.9            </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
