<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jean Morrison" />

<meta name="date" content="2020-05-07" />

<title>Simulating Data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">gwas_factors</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jean997/ss_factors">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Simulating Data</h1>
<h4 class="author">Jean Morrison</h4>
<h4 class="date">2020-05-07</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-05-11
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>sumstatFactors/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0.9000). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstreedd5b4eaf5f7ddf793d8c95992748c49c49fd864btargetblankdd5b4eaa"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jean997/sumstatFactors/tree/dd5b4eaf5f7ddf793d8c95992748c49c49fd864b" target="_blank">dd5b4ea</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjean997sumstatFactorstreedd5b4eaf5f7ddf793d8c95992748c49c49fd864btargetblankdd5b4eaa" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/fixed_factors_cache/
    Ignored:    analysis/metabo_flash_experiments_cache/
    Ignored:    analysis/pathway_factors_cache/
    Ignored:    analysis/simulations_cache/
    Ignored:    analysis/z_scores_cache/

Untracked files:
    Untracked:  R/est_R_pairwise.R
    Untracked:  R/sim_sumstats_lf.R
    Untracked:  analysis/2020-04-28-meeting_notes.Rmd
    Untracked:  analysis/genetic_correlation.Rmd
    Untracked:  analysis/ld_etc.Rmd
    Untracked:  analysis/sample_corr2.Rmd
    Untracked:  analysis/variant_selection.Rmd
    Untracked:  analysis/z_scores.Rmd
    Untracked:  buttons.css
    Untracked:  code/flashier_compare.R
    Untracked:  code/flashier_compare_jason.R
    Untracked:  code/hide_output.js
    Untracked:  code/mask_flashier.R
    Untracked:  code/mask_flashr.R
    Untracked:  covariance_scratch.R
    Untracked:  docs/figure/ld_etc.Rmd/
    Untracked:  docs/figure/sample_corr2.Rmd/
    Untracked:  k99_analysis.R
    Untracked:  k99_fit.RDS
    Untracked:  man/sim_sumstats_lf.Rd

Unstaged changes:
    Modified:   NAMESPACE

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jean997/sumstatFactors/blob/dd5b4eaf5f7ddf793d8c95992748c49c49fd864b/analysis/simulating_data.Rmd" target="_blank">dd5b4ea</a>
</td>
<td>
Jean Morrison
</td>
<td>
2020-05-12
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/simulating_data.Rmd”))
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="model" class="section level2">
<h2>Model</h2>
<p>Throughout <span class="math inline">\(m\)</span> indexes traits, <span class="math inline">\(j\)</span> indexes variants, <span class="math inline">\(k\)</span> indexes factors, and <span class="math inline">\(i\)</span> indexes individuals. We assume the following data generating model:</p>
<p><span class="math display">\[
Y_{m,i} = \sum_{k=1}^{K}f_{mk}F_{k,i} + \sum_{j=1}^{J} \theta_{j,m}G_{j,i} + \epsilon_{m,i}
\]</span> where <span class="math inline">\(F_{k,i}\)</span> is the value of factor <span class="math inline">\(k\)</span> for individual <span class="math inline">\(i\)</span>. <span class="math inline">\(\epsilon_{m,i}\)</span> is any non-genetic component of <span class="math inline">\(Y_{m,i}\)</span>. The covariance of <span class="math inline">\(\epsilon_{\bullet, i} = (\epsilon_{1,i}, \dots, \epsilon_{M,i})\)</span> is equal to <span class="math inline">\(\Sigma_E = S_{E}R_E S_E\)</span> where <span class="math inline">\(R_E\)</span> is the corresponding correlation matrix and <span class="math inline">\(S_E\)</span> is diagonal.</p>
<p>Each factor has a heritable and non-heritable component: <span class="math display">\[
F_{k,i} = \sum_{j=1}^J l_{j,k}G_{j,i} + \epsilon_{Fk}
\]</span> We assume that <span class="math inline">\(\epsilon_{Fk}\)</span> are mutually independent of each other and all the other random variables so far discussed.</p>
<p>We assume that <span class="math inline">\(\mathbf{Y}_m\)</span> and <span class="math inline">\(\mathbf{G}_{j}\)</span> are centered and scaled to have mean 0 and variance 1. For convenience, <span class="math inline">\(f_{m,k}\)</span> are scaled so that <span class="math inline">\(\sum_{m=1}^M f_{m,k}^2 =1\)</span> for all <span class="math inline">\(k\)</span>.</p>
<p>The marginal (and also joint because SNPs are independent) association of <span class="math inline">\(G_j\)</span> with <span class="math inline">\(Y_m\)</span> is <span class="math inline">\(\beta_{j,m} = \sum_{k=1}^{K}l_{jk}f_{mk} + \theta_{j,m}\)</span> (this is our basic factor decomposition equation). We view <span class="math inline">\(\theta_{j,m}\)</span> and <span class="math inline">\(l_{j,k}\)</span> as random variables that are independent of each other and are mutually independent across SNPs and traits and mean 0. The GWAS estimate of <span class="math inline">\(\beta_{j,m}\)</span> is</p>
<p><span class="math display">\[
\hat{\beta}_{j,m} = \frac{1}{N_m}\mathbf{G}_j^\top \mathbf{Y}_{m}
\]</span> where <span class="math inline">\(N_m\)</span> is the sample size for trait <span class="math inline">\(m\)</span>. <span class="math inline">\(Var(\hat{\beta}_{j,m}) \approx 1/N_m\)</span>, assuming no variant makes a large contribution to the variance of <span class="math inline">\(Y_m\)</span>.</p>
<p>If GWAS for all traits have the same sample (<span class="math inline">\(N_m \equiv N\)</span>) and no individual variant makes a large contribution to heritability, covariance of <span class="math inline">\(\hat{\beta}_{j, \bullet}\)</span> is approximately the trait covariance scaled by <span class="math inline">\(1/N\)</span>. Denote trait covariance <span class="math inline">\(\Sigma\)</span>.</p>
<p><span class="math display">\[
\Sigma = \Sigma_G + \Sigma_{FE} + \Sigma_E
\]</span> where <span class="math inline">\(\Sigma_G\)</span> is the genetic covariance, <span class="math inline">\(\Sigma_{FE}\)</span> is the non-genetic covariance due to shared factors and <span class="math inline">\(\Sigma_E\)</span> is environmental covariance not mediated by the heritble factors.</p>
<p><span class="math display">\[
(\Sigma_G)_{m, m^\prime} = \sum_{j=1}^J E\left[\left(\sum_{k=1}^K l_{jk}f_{mk} + \theta_{jm}\right)\left(\sum_{k=1}^K l_{jk}f_{m^\prime k} + \theta_{jm^\prime}\right)\right]  \\
 = \sum_{j=1}^JE\left[\sum_{k =1}^{K} l_{j,k}^2 f_{m,k} f_{m^\prime, k}\right]  + E\left[\theta_{j,m}\theta_{j,m^\prime}\right]   \\
 = J \sum_{k=1}^K \sigma_{lk}^2  f_{m,k} f_{m^\prime, k} + JCov(\theta_m, \theta_{m^\prime})
 \]</span> Where <span class="math inline">\(\sigma_{lk}^2 = Var(l_{j,k})\)</span> and <span class="math inline">\(Cov(\theta_m, \theta_{m^\prime}) = \sigma_{\theta,m}^2\)</span> if <span class="math inline">\(m = m^\prime\)</span> and 0 otherwise. In matrix form</p>
<p><span class="math display">\[
\Sigma_G = J (FV_LF^\top + V_\theta)
\]</span> where <span class="math inline">\(V_\theta\)</span> is a diagonal <span class="math inline">\(M\times M\)</span> matrix who’s elements are equal to <span class="math inline">\(\sigma^2_{\theta, m}\)</span> and <span class="math inline">\(V_L\)</span> is diagonal <span class="math inline">\(K\times K\)</span> with elements <span class="math inline">\(\sigma^2_{l,k}\)</span>.</p>
<p><span class="math inline">\(\Sigma_{FE} = FV_{F}F^\top\)</span> where <span class="math inline">\(V_F\)</span> is a <span class="math inline">\(K \times K\)</span> diagonal matrix who’s elements are <span class="math inline">\(Var(\epsilon_{Fk}) \equiv \sigma^2_{Fk}\)</span>.</p>
</div>
<div id="simulating-data" class="section level2">
<h2>Simulating Data</h2>
<p>Parameters specified to define simulation:</p>
<ul>
<li><span class="math inline">\(N\)</span>: GWAS sample size</li>
<li><span class="math inline">\(h^2_1, \dots, h^2_{M}\)</span>: Heritability of each trait.</li>
<li><span class="math inline">\(\Omega_m\)</span>: Proportion of SNP heritability mediated by factors.</li>
<li><span class="math inline">\(d^2_1, \dots, d^2_K\)</span>: heritability of each factor.</li>
<li><span class="math inline">\(\pi_1, \dots, \pi_K\)</span>: SNP sparsity of each factor.</li>
<li><span class="math inline">\(R_E\)</span>: Environmental trait correlation not mediated by factors.</li>
<li><span class="math inline">\(J\)</span>: Total number of SNPs</li>
</ul>
<p>Simulation procedure:</p>
<ol style="list-style-type: decimal">
<li>Generate <span class="math inline">\(L = (l_{jk})\)</span></li>
<li>Generate <span class="math inline">\(\Theta = (\theta_{j,m})\)</span></li>
<li>Compute <span class="math inline">\(\beta = (\beta_{j,m}) = LF^\top + \theta\)</span></li>
<li>Compute row covariance of error, <span class="math inline">\(\frac{1}{N}\Sigma\)</span>.</li>
<li>For each <span class="math inline">\(j\)</span>, sample <span class="math inline">\(E_{j,\bullet}\)</span> from a <span class="math inline">\(N_{M}(0, \frac{1}{N}\Sigma)\)</span> distribution</li>
<li>Compute <span class="math inline">\(\hat{\beta}_{j,\bullet} = \beta_{j,\bullet} + E_{j, \bullet}\)</span></li>
<li>The standard error of <span class="math inline">\(\hat{\beta}_{j, \bullet}\)</span> is <span class="math inline">\(\sqrt{(sqrt(\Sigma)/N}\)</span></li>
</ol>
<p>Without constraining the model, we can set <span class="math inline">\(\sigma_{l,k} = 1/J\)</span> for all <span class="math inline">\(k\)</span> making <span class="math inline">\(f_{m,k}^2\)</span> interpretable as the proportion of variance contributed to trait <span class="math inline">\(m\)</span> by the heritable component of factor <span class="math inline">\(k\)</span>. This requires re-scaling rows of <span class="math inline">\(F\)</span> so that</p>
<p><span class="math display">\[
\sum_{k=1}^{K}f_{m,k}^2 = \Omega_m h^2_m
\]</span></p>
<p>We need to determine <span class="math inline">\(\sigma_{\theta, m}\)</span>, <span class="math inline">\(\sigma_{F,k}\)</span>, <span class="math inline">\(\sigma^2_{E,m}\)</span> in terms of the specified model parameters.</p>
<p>We have the following relationships for each <span class="math inline">\(m\)</span> and <span class="math inline">\(k\)</span></p>
<p><span class="math display">\[
1 = h^2 + \sum_{k=1}^{K}\sigma^2_{Fk}f_{m,k}^2 + \sigma^2_{E,m}\\
(1-\Omega_m)h^2_m =  J\sigma^2_{\theta,m} \\
d^2_k =  \frac{1}{1 + \sigma^2_{Fk}} \\
\]</span> This specifies all of the necessary parameters:</p>
<p><span class="math display">\[
\sigma_{Fk}^2 = \frac{1-d^2_k}{d^2_k}\\
\sigma^2_{\theta,m} = \frac{1}{J} (1-\Omega_m)h^2_m\\
\sigma_{E,m}^2 = 1 - h^2 - \sum_{k=1}^K \frac{1-d^2_k}{d^2_k}f_{m,k}^2
\]</span> Some notes about the impact of these parameters:</p>
</div>
<div id="adding-ld" class="section level2">
<h2>Adding LD</h2>
<p>We can add LD by performing the following steps given a SNP correlation matrix <span class="math inline">\(R_{LD} = U_{LD}D_{LD}U_{LD}^\top\)</span>:</p>
<ol style="list-style-type: decimal">
<li><p>Convert to z-scores: <span class="math inline">\(Z = \beta/S\)</span> where <span class="math inline">\(S\)</span> is the <span class="math inline">\(J \times M\)</span> matrix of standard errors and we are using “R notation” to just mean element-wise division of matrices. Compute <span class="math inline">\(E_z = E/S\)</span> where <span class="math inline">\(E\)</span> is the matrix of errors with row correlation that we generated in the previous section.</p></li>
<li><p>Compute $ = R_{LD}Z + U_{LD}D^{1/2}E_{Z} $</p></li>
<li><p>Convert back to effect sizes if desired.</p></li>
</ol>
<p>Of note: This procedure generates summary statisitcs with the correct row and column covariance. I am currently unsure if knowing the row and column covariance of a matrix uniquely determines the full covariance of the vectorized matrix (i.e. <span class="math inline">\(Cov(E_{j,m}, E_{j^\prim, m^\prime})\)</span>, <span class="math inline">\(j\neq j^\prime\)</span>, <span class="math inline">\(m \neq m^\prime\)</span>).</p>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
