---
title: "Summary Statistics Factors Paper Outline and To Do"
author: "Jean Morrison"
date: "August 10, 2020"
output: pdf_document
---

### Summary


The overall thesis of the paper is simple. We are presenting a strategy which is FLASH + sample overlap correction for estimating factors from largish sets of summary statistics (lets say 10 to a few hundred). We are arguing that this strategy is better at recovering the "true" factors than a matrix decomposition that requires factors to be orthogonal. The main alternative now I think is DeGas (Tanigawa, Nature Communications, 2019) which uses truncated SVD applied to a modified matrix of summary statistics in which many effects are shrunk to zero. We want to demonstrate in simulations and data application that FLASH is better than DeGas (or similar orthogonal methods) with "better" defined by:
Simulations: True factors more accurately recovered
GWAS Data: Factors recovered are more interpretable in terms of function

Ideally we can also offer some kind of useful pipeline. It would be nice for example to go from a list of IEU GWAS codes to output fairly simply.

To Do: Do we need a name for the strategy?

### Outline

#### Introduction: 
  
  + Introduce the problem
  +  why one would want to decompose summary statistics
  + how FLASH with sparse priors differs from orthogonal decompositions
  + shortcomings of existing methods: I don't think DeGas accounts for sample overlap. I should double check this.
  

#### Methods
  
  1.  Introduce model: This is fairly simple. I think the clearest place I have written out all the parts is [here](simulating_data.html) where I wrote about how to simulate data from the model. Here we would just introduce the simplest parts and save LD and sample overlap for later sections.
  2. Data harmonization and effect normalization. 
    Effect normalization: We want to do some normalization of effect estimates across traits. Ideally we would like to say that if factor k explains a% of the variance of trait m, that is equivalent to saying factor k explains a% of the heritability of trait m that is explained by the full set of variants include. I am currently using $(\hat{\beta}_{j,m}/SE_{j,m})*\sqrt{N_m}$. 
  3. Sample overlap: Demonstrate the effects of sample overlap on the correlation between summary statistics. Introduce methods for correcting for sample overlap. Currently we have 
  
  + Fixed factors (FF): Add fixed factors with specified priors. Fixed factors are eigenvectors of an estimated correlation matrix.
  + Eigenvector transformation  (EV)
  
  EV is more accurate than FF in simulations. It is more of an "exact" correction in that FF involves some estimation of the variance of the fixed factors which might have an effect of overcorrecting if the correlation due to sample overlap is similar to the genetic correlation which may often be the case. EV can only be used with complete data which is a fairly big limitation. It also imposes the sparsity assumption on the wrong matrix which may be a problem if the correlation due to sample overlap is not itself sparse. In real data I have mixed results about which seems to work better but that is complicated by my opinion of what "better" looks like in real data. 
  
  Estimation of R, the correlation due to sample overlap, is an outstanding issue. In my simulations, it seems that estimating R with all variants is better than using only non-significant variants. However, I am not sure this is the case in the GWAS data, in particular when we don't know the extent of sample overlap we can end up overestimating the correlation due to sample overlap and overcorrecting for it.
  
  4. Uncorrected population structure. Uncorrected population structure should create a dense factor. The fixed factor correction method plus an agnostic approach to estimating R (one that does not use information about which studies have overlapping samples) should remove all dense factors so I think this may also remove "population structure factors". The other strategy for checking for population structure factors is to check if SNP loadings correlate with population PCs.
  
  5. Linkage disequilibrium: We deal with it by pruning.
  
#### Simulations

In simulations mostly we want to show that FLASH is recovering factors more accurately than orthogonal methods.

Main simulations:

+ Independent SNPs
+ Look at both K > M and K < M scenarios. 
+ Vary distribution of proportion of variance explained by each factor. The hardest setting is probably K >> M and uniform distribution of per-factor PVE.

Additional:

+ Demonstrate effect of matching and mismatching LD
+ Demonstrate robustness to different patterns of missingness

Current: I have only run small simulations to test EV and FF and to test different strategies for estimating R. I have written some new code that makes it possible to specify more things, like the PVE for each factor and the sparsity of the error matrix.

#### Data
  A major challenge here is that running these application requires a significant amount of curation. This would be a great thing to get another person involved on. There is alos a project of creating a nice usable pipeline that still allows for data from different sources to be used.
  
  Some possible applications to present: 
  
  1. Immune mediated and psychiatric traits: Ideally we can connect our results with results about immune mediated traits. There are several papers describing the structure of these and there is also Zhu et al, European Respiratory Journal, 2019 which talks about overlap of asthma and mental health disorders. Ideally we can show another angle on this problem. I am currently running one iteration of this analysis but we may go back and make different analysis choices after resolving some of the issues above.
  
  2. Different architectures of anthropometric traits in males and females: We can use the Neale lab UK biobank results here which have versions separated by sex. I haven't run this yet to see if there are interesting results but I expect there to be some. Khramtsova et al, NRG, 2019 is a good review of sex differences in genomics.


