---
title: "Immune Mediated Diseases"
author: "Jean Morrison"
date: "2020-06-18"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction

This is an exploration of an analysis of a set of immune mediated diseases exploring some analysis choices. 

Disease |Abbrv |  Cases | Controls | SNPS  
--------|--------|--------|----------|------
Lupus | SLE    |  7219  | 15991    | 6,762,880
Rheumatoid Arthritis | RA     | 29880  | 73758    |  8,240,308 
Chrons Disease| CD     | 22575	| 46693    | 9,458,484 
Inflammatory Bowel  Disease | IBD    | 42950	| 53536    | 9,934,092
Ulcerative Colitis| UC     | 20417	| 52230    |  9,554,059 
Asthma | Asth   | 23948  |	118538   | 2,001,281 
Primary Schlerosing Cholangitis | PSC    | 4796	  | 19955    | 6,805,391 
Allergy | Allg   | 180129	| 180709   | 7,112,974 
Ankylosing Spondilitis | Anksp  | 10619	| 15145    | 103,085 
Celiac | Cel    | 4533	  | 10750    | 521,934 
Multiple Sclerosis| MS     | 14498	| 24091    | 136,408 
Psoriasis| PSO    | 10588	| 22806    | 141,528 
Primary Biliary Cirrhosis | PBC    |2764	  | 10475    | 1,047,436 
hildhood Ashtma | Ch. Asth (Ukb) | 9676 |	347481 | 7,038,206 
Ashtma (both) | Asth (Ukb) |46802 | 	347481 | 7,038,206 
Adult onset asthma | Ad. Asth (Ukb) | 22296 |	347481 | 7,038,206 
Latend autoimmune diabetes | LADA | 2634	| 5947 | 4,293,816 
Vitiligo | Vit | 2853	| 37405 | 7,490,324
Type 1 Diabetes | T1D | 6683	| 12173 |  107,814
 
There are a few challenges with these data. The fist is that some but not all studies have overlapping samples. UC, CD, and IBD are all measured in the same samples as are three of the asthma phenotypes (the three marked as (Ukb) above). A second challenge is that some the othe studies, Anksp, MS, PSO, and T1D only have published summary statistics for immunochip SNPs which is a much smaller set than the usual genome-wide study. 
In all analyes variants are pruned for LD

## Non-missing variants, various sample overlap corrections

The first set of analyses uses only variants that are non-missing in all studies and uses several methods to correct for sample overlap. we consider three estimates of $R$, the residual correlation in rows of $E$.

1. Identity (no sample overlap)
2. $R$ estimated with all SNPs in all studies (assumes possible sample overlap between any pair of studies) ("all").
3. Correlation estimated only between studies with known overlapping samples ("overlap"). 

We consider two methods of correcting for sample overlap, the eigenvector transformation and fixed factors. In this case, since studies are mostly non-overlapping this analysis tells us how much power/accuracy we lose if we correct for sample overlap when it is not present. For comparison we will also look at the SVD of all variants.

```{r, echo=FALSE}
library(sumstatFactors)
library(ggplot2)
library(dplyr)
library(gridExtra)

res_none <- readRDS("analysis_data/2020-06-10_ai/ai_fit_Inf_100_naive_TRUE_2.RDS")
p1 <- with(res_none, plot_factors(F_hat, names)) + ggtitle("non-missing, R = I")

res_ev1 <- readRDS("analysis_data/2020-06-10_ai/ai_fit_Inf_100_ev_TRUE_2.RDS")
p_ev1 <- with(res_ev1, plot_factors(t(t(F_hat)/sqrt(colSums(F_hat^2))), names)) + ggtitle("non-missing, R1 (all), ev")

res_ff1 <- readRDS("analysis_data/2020-06-10_ai/ai_fit_Inf_100_ff_TRUE_2.RDS")
p_ff1 <- with(res_ff1, plot_factors(t(t(F_hat)/sqrt(colSums(F_hat^2))), names)) + ggtitle("non-missing, R1 (all), ff")

res_ev2 <- readRDS("analysis_data/2020-06-10_ai/ai_fit_Inf_100_ev2_TRUE_2.RDS")
p_ev2 <- with(res_ev2,plot_factors(t(t(F_hat)/sqrt(colSums(F_hat^2))), names)) + ggtitle("non-missing, R2 (overlap), ev")

res_ff2 <- readRDS("analysis_data/2020-06-10_ai/ai_fit_Inf_100_ff2_TRUE_2.RDS")
p_ff2 <- with(res_ff2, plot_factors(t(t(F_hat)/sqrt(colSums(F_hat^2))), names))+ ggtitle("non-missing, R2 (overlap), ff")

```

Without correcting for sample overlap:
```{r, echo=F}
p1
```

A few interesting patterns: factors 1 and 2 capture IBD/UC/CD and ashthma/allergy clusters respectively. These are probably the strongest patterns in these data. Factor 5 captures differences between UC and CD plus some overlap of UC and PSC and RA which can be concommitant. Factor 6 is unique to childhood (not adult onset) asthma and is also shared with PBC, SLE, and RA. 

Correcting for sample overlap across all studies:
```{r, echo=F}
grid.arrange(p_ev1, p_ff1, ncol=2)
```

Results using the eigenvector transformation are "muddy" while the fixed factor estimator is more sparse and has similar patterns to the using $R=I$. Factor 12 captures sharing between T1D and LADA which we would expect to see.

Correcting for sample overlap in only some studies:
```{r, echo=FALSE}
grid.arrange(p_ev2, p_ff2, ncol=2)
```

Here are two SVD decompositions. In the first we just take the SVD of the entire matrix of $z$-scores. In the second, $z$-scores corresponding to $p$-values greater than 0.05 are set to zero. These are extremely similar.
```{r, echo=F}
svd_res <-  readRDS("analysis_data/2020-06-10_ai/ai_svd.RDS")
plot_factors(svd_res$svd1$v, res_ev2$names) + ggtitle("SVD all")
plot_factors(svd_res$svd2$v, res_ev2$names) + ggtitle("SVD sparse")

```

## Variants present in at least two studies

If we allow missing data we can use many more variants. This analysis has 168,142 variants rather than 3,991. We can no longer use the eigenvector transformation. Both with and without correction, these fits are "less neat" than the non-missing data fits. This may because the imuno chip traits have very high (90%) missingness. To explore further, I will analyze only the non-imunochip traits.

```{r, echo=FALSE}
res_none <- readRDS("analysis_data/2020-06-10_ai/ai_fit_Inf_100_naive_FALSE_2.RDS")
p1 <- with(res_none, plot_factors(F_hat, names)) + ggtitle("allow missing, R = I")

res_ff1 <- readRDS("analysis_data/2020-06-10_ai/ai_fit_Inf_100_ff_FALSE_2.RDS")
p_ff1 <- with(res_ff1, plot_factors(t(t(F_hat)/sqrt(colSums(F_hat^2))), names)) + ggtitle("allow missing, R1 (all), ff")

p1
p_ff1

```

## Non-missing variants, non immunochip studies
